{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<html>\n",
    "    <summary></summary>\n",
    "    <p float=\"left\">\n",
    "         <div> <p></p> </div>\n",
    "         <div style=\"font-size: 20px; width: 800px;\">\n",
    "              <h1>\n",
    "               <left>Importing packages, classes/modules, os navigation, reading files </left>\n",
    "              </h1>\n",
    "              <p><left>============================================================================</left> </p>              \n",
    "             <pre>Course: BIOM/CBE 480A5, Spring 2025\n",
    "Instructor: Dr. Brian Munsky\n",
    "Contact Info: munsky@colostate.edu\n",
    "Authors: Will Raymond, Dr. Brian Munsky\n",
    "</pre>\n",
    "         </div>\n",
    "    </p>\n",
    "\n",
    "</html>\n",
    "\n",
    "\n",
    "\n",
    "<details>\n",
    "  <summary>Copyright info</summary>\n",
    "\n",
    "```\n",
    "Copyright 2023 Brian Munsky\n",
    "\n",
    "Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:\n",
    "\n",
    "1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.\n",
    "\n",
    "2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.\n",
    "\n",
    "3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.\n",
    "\n",
    "THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n",
    "```\n",
    "<details>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_YPwTneNwC64"
   },
   "source": [
    "# **Learning Objectives**\n",
    "In this notebook, we will learn about one of the nicest things about python -- the modular nature of object oriented programming. After completing this notebook, you will be able to:\n",
    "* ```import``` packages or modules to use in your codes or notebooks\n",
    "* find and ```install``` new packages\n",
    "* run simple ```os``` (operating system) tasks to control your computer\n",
    "* ```open``` and ```read/write``` to files\n",
    "* work with ```pandas``` to manage files and data\n",
    "* save data using ```numpy``` arrays and files\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **1) Importing, Using, and Installing packages**\n",
    "\n",
    "```Reading: Kinder, Nelson Section 4 and Appendix B```\n",
    "\n",
    "Importing modules lets you load particular code into your instance of Python and use only the ones needed.  \n",
    "\n",
    "For example, we can import one section or function of a given code instead of loading the entire package. Available modules are based on your current path / environment variables and your current instillation of python.\n",
    "\n",
    "You can import the package ```sys``` (system) and take a glance at what paths (folders/directories) python has access to when importing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "-g7iivNAb5Q7",
    "outputId": "afe14cb2-9884-432f-f19e-09c6881455f6",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\python310.zip',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\DLLs',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5',\n",
       " '',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\win32',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\win32\\\\lib',\n",
       " 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\Pythonwin']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import the modules os and sys\n",
    "import os, sys\n",
    "\n",
    "# Print the current path using the function path from the module sys.  The result of this function\n",
    "# is a list of strings, each string representing a path.  The first path is the current path, which\n",
    "# is the path where the script is being executed. The other paths are the paths where python will\n",
    "# look for modules to import.  If you are trying to import a module and you get an error, you can\n",
    "# check to make sure that the path to the module is in this list.\n",
    "sys.path  #current paths python has"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0L49iqF6J3gI"
   },
   "source": [
    "Let's take a glance at what we have installed already (when we created our environment). To do this we are going to call ```pip```, ```pip``` is the main package manager for Python and comes included since Python 3. ```pip```connects to [PyPI](https://pypi.org/), the python package index where people are free to release code and modules that they have developed.\n",
    "\n",
    "**Tip - The % symbol below tells VS Code to run this command in the background unix kernel, as if this were a terminal or command prompt input.** By calling ```!pip list``` or ```%pip list``` we are telling the terminal to list every package.  Alternatively, try just running the command ```pip list''' in your terminal, and you will get the same result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "U9pWqC3qKG94",
    "outputId": "6b622281-dd29-4494-d4fb-38b5ac066793",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Run a \"magic\" command to list the modules that are currently installed in the python environment.\n",
    "# Here the symbol \"%\" is used to indicate that the command is a magic command.  The command \"pip\"\n",
    "# is used to install python packages.  The command \"list\" is used to list the packages that are\n",
    "# currently installed.  The result of this command is a list of strings, each string representing\n",
    "# a package that is installed.  If you are trying to import a module and you get an error, you can\n",
    "# check to make sure that the package is in this list.\n",
    "%pip list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Ac061_wIM64K"
   },
   "source": [
    "To get information about any of these modules, we need to import it and then use ```!module?``` or ```help(module)``` to get additional information about the package. What is displayed is from a specifically formatted comment called a ```docstring``` (documentation string) that comes along with the package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "MOBgOvXAMt9x",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;31mType:\u001b[0m        module\n",
      "\u001b[1;31mString form:\u001b[0m <module 'numpy' from 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\numpy\\\\__init__.py'>\n",
      "\u001b[1;31mFile:\u001b[0m        c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages\\numpy\\__init__.py\n",
      "\u001b[1;31mDocstring:\u001b[0m  \n",
      "NumPy\n",
      "=====\n",
      "\n",
      "Provides\n",
      "  1. An array object of arbitrary homogeneous items\n",
      "  2. Fast mathematical operations over arrays\n",
      "  3. Linear Algebra, Fourier Transforms, Random Number Generation\n",
      "\n",
      "How to use the documentation\n",
      "----------------------------\n",
      "Documentation is available in two forms: docstrings provided\n",
      "with the code, and a loose standing reference guide, available from\n",
      "`the NumPy homepage <https://numpy.org>`_.\n",
      "\n",
      "We recommend exploring the docstrings using\n",
      "`IPython <https://ipython.org>`_, an advanced Python shell with\n",
      "TAB-completion and introspection capabilities.  See below for further\n",
      "instructions.\n",
      "\n",
      "The docstring examples assume that `numpy` has been imported as ``np``::\n",
      "\n",
      "  >>> import numpy as np\n",
      "\n",
      "Code snippets are indicated by three greater-than signs::\n",
      "\n",
      "  >>> x = 42\n",
      "  >>> x = x + 1\n",
      "\n",
      "Use the built-in ``help`` function to view a function's docstring::\n",
      "\n",
      "  >>> help(np.sort)\n",
      "  ... # doctest: +SKIP\n",
      "\n",
      "For some objects, ``np.info(obj)`` may provide additional help.  This is\n",
      "particularly true if you see the line \"Help on ufunc object:\" at the top\n",
      "of the help() page.  Ufuncs are implemented in C, not Python, for speed.\n",
      "The native Python help() does not know how to view their help, but our\n",
      "np.info() function does.\n",
      "\n",
      "To search for documents containing a keyword, do::\n",
      "\n",
      "  >>> np.lookfor('keyword')\n",
      "  ... # doctest: +SKIP\n",
      "\n",
      "General-purpose documents like a glossary and help on the basic concepts\n",
      "of numpy are available under the ``doc`` sub-module::\n",
      "\n",
      "  >>> from numpy import doc\n",
      "  >>> help(doc)\n",
      "  ... # doctest: +SKIP\n",
      "\n",
      "Available subpackages\n",
      "---------------------\n",
      "lib\n",
      "    Basic functions used by several sub-packages.\n",
      "random\n",
      "    Core Random Tools\n",
      "linalg\n",
      "    Core Linear Algebra Tools\n",
      "fft\n",
      "    Core FFT routines\n",
      "polynomial\n",
      "    Polynomial tools\n",
      "testing\n",
      "    NumPy testing tools\n",
      "distutils\n",
      "    Enhancements to distutils with support for\n",
      "    Fortran compilers support and more  (for Python <= 3.11).\n",
      "\n",
      "Utilities\n",
      "---------\n",
      "test\n",
      "    Run numpy unittests\n",
      "show_config\n",
      "    Show numpy build configuration\n",
      "matlib\n",
      "    Make everything matrices.\n",
      "__version__\n",
      "    NumPy version string\n",
      "\n",
      "Viewing documentation using IPython\n",
      "-----------------------------------\n",
      "\n",
      "Start IPython and import `numpy` usually under the alias ``np``: `import\n",
      "numpy as np`.  Then, directly past or use the ``%cpaste`` magic to paste\n",
      "examples into the shell.  To see which functions are available in `numpy`,\n",
      "type ``np.<TAB>`` (where ``<TAB>`` refers to the TAB key), or use\n",
      "``np.*cos*?<ENTER>`` (where ``<ENTER>`` refers to the ENTER key) to narrow\n",
      "down the list.  To view the docstring for a function, use\n",
      "``np.cos?<ENTER>`` (to view the docstring) and ``np.cos??<ENTER>`` (to view\n",
      "the source code).\n",
      "\n",
      "Copies vs. in-place operation\n",
      "-----------------------------\n",
      "Most of the functions in `numpy` return a copy of the array argument\n",
      "(e.g., `np.sort`).  In-place versions of these functions are often\n",
      "available as array methods, i.e. ``x = np.array([1,2,3]); x.sort()``.\n",
      "Exceptions to this rule are documented."
     ]
    }
   ],
   "source": [
    "# Let's get some help on the numpy module.  We can import the module and then use the help command\n",
    "import numpy\n",
    "# If we want to learn more about a module, we can use the help command !(name)?\n",
    "!numpy?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Or we can use the help function:\n",
    "help(numpy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numpy?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_gTByrOwWrD8"
   },
   "source": [
    "### **1.A) importing different packages or sub-modules**\n",
    "\n",
    "```import``` in Python is the command to load in a particular ```package``` (large collection of code with several submodules) or ```module``` (single files up to medium sized classes).\n",
    "\n",
    "When you import you have several options:\n",
    "\n",
    "```python\n",
    "# import the entire package all at once\n",
    "import numpy\n",
    "\n",
    "# import the package and use a (simpler) variable name, \"np\" to refer to it\n",
    "import numpy as np\n",
    "\n",
    "# import a specific submodule or function from the package\n",
    "from numpy import linalg as la\n",
    "\n",
    "# importing one function only from the sub - module linear algebra contained in numpy\n",
    "from numpy.linalg import svd as svd\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 36
    },
    "id": "bXXO_iNVvbqP",
    "outputId": "9ae649ce-6979-4a5f-c02f-8728eb4a3790",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.26.4'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# It is very common when using numpy to import it with an alias.  The alias is a shorter name that\n",
    "# can be used to refer to the module.  The alias is specified using the \"as\" keyword.  Here we import\n",
    "# numpy with the alias \"np\".  This is a very common alias for numpy.\n",
    "import numpy as np\n",
    "\n",
    "# if we want to know the version:\n",
    "np.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 36
    },
    "id": "TNCoa_yYWqdz",
    "outputId": "8ad54838-fd35-4c74-90f3-65eaf4a71c1a",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\numpy\\\\linalg\\\\__init__.py'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If we only want to import a specific function from a module, we can do that.  Here we import the\n",
    "# function \"linalg\" from the module \"numpy\".  This is useful if we only need a specific function\n",
    "# from a module and we don't want to import the entire module.\n",
    "from numpy import linalg as la #import only the linear algebra file\n",
    "# or if we want to know where a specific file is:\n",
    "la.__file__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# or if we want just the docstring:\n",
    "la.__doc__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "FT5tUqRVYAOl",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Let's import the SVD function from the numpy module and use the help command to learn more about it.\n",
    "from numpy.linalg import svd as svd\n",
    "svd?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Zhfov8rCmBan",
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### **Importing local packages**\n",
    "You can also write your own packages and then import them, but the folders need to be written in the right format, and the folders need to be in the search path.\n",
    "\n",
    "To learn more, see the page \"Writing your own callable functions and packages\" on canvas, but here is an example that we have used already. Let's look at what is in the folder \"../BIOM480A5\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's look at the contents of the directory \"../BIOM480A5\".  We can use the listdir function from\n",
    "# the os module to list the contents of a directory.  The argument to the listdir function is the\n",
    "# path to the directory.  Here we use the function path from the os module to get the absolute path\n",
    "# to the directory \"../BIOM480A5\".  The result of the listdir function is a list of strings, each\n",
    "# string representing a file or directory in the directory.\n",
    "try:\n",
    "    os.listdir(os.path.abspath('../BIOM480A5'))\n",
    "    BIOM480A5_path = os.path.abspath('../BIOM480A5')\n",
    "except:\n",
    "    os.listdir(os.path.abspath('BIOM480A5'))\n",
    "    BIOM480A5_path = os.path.abspath('BIOM480A5')\n",
    "# Note: The try and except statements are used to catch errors.  If an error occurs in the try block,\n",
    "# then the except block is executed.  This is useful for catching errors that may occur when running\n",
    "# code.  Here we use the try and except statements to catch an error that may occur if the path\n",
    "# \"../BIOM480A5\" does not exist because we open the notebook from a different directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice, the file ```\"__init__.py ``` in the folder.  This is an empty file that is used to indicate that the directory is a python package.\n",
    "\n",
    "To import the module BIOM480A5, we need to make sure that this folder is in the search path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the path to the directory \"../BIOM480A5\" to the list of paths that python will look for modules\n",
    "# to import.  This is useful if you want to import a module from a directory that is not in the default\n",
    "# list of paths.\n",
    "sys.path.append(BIOM480A5_path)\n",
    "\n",
    "# now we can import the module \"BIOM480A5_Quizzes\" from the directory \"../BIOM480A5\". \n",
    "import BIOM480A5_Quizzes\n",
    "\n",
    "# Or we can import a specific function from the module\n",
    "from BIOM480A5_Quizzes import multiple_choice_quiz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Refreshing import to get changes to local libraries**\n",
    "\n",
    "If you make changes to a local library, you will need to reload your imported functions in order to use the new codes.  Here is an example for how to do this: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'BIOM480A5_Quizzes' from 'c:\\\\Users\\\\histe\\\\OneDrive\\\\Documents\\\\GitHub\\\\480A5CourseMaterials\\\\BIOM480A5\\\\BIOM480A5_Quizzes.py'>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import importlib\n",
    "\n",
    "# If we make changes to the module \"BIOM480A5_Quizzes\" and we want to reload the module, we can use\n",
    "importlib.reload(BIOM480A5_Quizzes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "F3BLTyKhlX7l",
    "tags": []
   },
   "source": [
    "### **1.B) Using packages**\n",
    "\n",
    "Once we have imported a package, you can call any ```submodules``` or functions by ```module.function()```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "kXotxOskV2TO",
    "outputId": "14037dc8-6096-46c6-8ca5-ad6d4d5bb5c8",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Let's use the numpy and SVD functions that we imported to calculate the SVD of a matrix.  We will\n",
    "# create a matrix using the numpy function \"ones\" and then use the SVD function to calculate the\n",
    "# singular value decomposition of the matrix.\n",
    "arr = np.ones([10,100])\n",
    "svd_results = la.svd(arr)\n",
    "\n",
    "# The result of the SVD function is a tuple of three arrays.  The first array is the U matrix, the\n",
    "# second array is the singular values, and the third array is the V matrix.  We can access these\n",
    "# arrays by indexing the tuple.\n",
    "U = svd_results[0]\n",
    "S = svd_results[1]\n",
    "V = svd_results[2]\n",
    "\n",
    "# Let's print the shapes (i.e., sizes) of the U, S, and V arrays.\n",
    "print(U.shape)\n",
    "print(S.shape)\n",
    "print(V.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PKdiuk7rlSUJ"
   },
   "source": [
    "### **1.C) Installing new packages**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "64Q6SW3jNZmp"
   },
   "source": [
    "To find more packages to install, usually they are stored on [PyPI](https://pypi.org/) - Python Package Index. \n",
    "\n",
    "Because there are so many user supplied packages, searching for specific modules can be difficult difficult within PyPI, so we recommend finding packages via searching academic papers or Google scholar for your specific purposes and using the authors' particular Githubs!\n",
    "\n",
    "From there, you can install a package into your environment by using ```!pip``` according to the package's instructions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "-5hjsl9LNyvV",
    "outputId": "b1cb29da-bffe-4d94-c543-25888ce4febe",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'skbio'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 8\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Let's try to install the package scikit-bio.  This package is not installed by default in the\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;66;03m# python environment, so we will need to install it.  We can use the pip command to install the\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;66;03m# package.  The command is \"pip install\" followed by the name of the package.  The exclamation\u001b[39;00m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;66;03m# point is used to indicate that the command is a shell command.  \u001b[39;00m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;66;03m# https://scikit.bio\u001b[39;00m\n\u001b[0;32m      7\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpip install scikit-bio\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m----> 8\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mskbio\u001b[39;00m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m     10\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe version of scikit-bio installed:\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'skbio'"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for scikit-bio (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [836 lines of output]\n",
      "      [1/7] Cythonizing skbio/metadata/_intersection.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\metadata\\_intersection.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      [2/7] Cythonizing skbio/stats/ordination/_cutils.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\stats\\ordination\\_cutils.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      [3/7] Cythonizing skbio/tree/_c_nj.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\tree\\_c_nj.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      [4/7] Cythonizing skbio/alignment/_ssw_wrapper.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\alignment\\_ssw_wrapper.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      [5/7] Cythonizing skbio/diversity/_phylogenetic.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\diversity\\_phylogenetic.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      [6/7] Cythonizing skbio/stats/distance/_cutils.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\Cython\\Compiler\\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\\Users\\histe\\AppData\\Local\\Temp\\pip-install-1o1ubvrd\\scikit-bio_7cd7bca9d17c48279a3fffaaff831ded\\skbio\\stats\\distance\\_cutils.pyx\n",
      "        tree = Parsing.p_module(s, pxd, full_module_name)\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:398:5: Exception check on '_dist_euclidean' will always require the GIL to be acquired. Declare '_dist_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:407:5: Exception check on '_norm_euclidean' will always require the GIL to be acquired. Declare '_norm_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:414:5: Exception check on '_sum' will always require the GIL to be acquired. Declare '_sum' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:350:36: Exception check after calling '__pyx_fuse_0_dist_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_0_dist_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:357:24: Exception check after calling '__pyx_fuse_0_sum' will always require the GIL to be acquired. Declare '__pyx_fuse_0_sum' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:381:35: Exception check after calling '__pyx_fuse_0_norm_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_0_norm_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:389:34: Exception check after calling '__pyx_fuse_0_dist_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_0_dist_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:350:36: Exception check after calling '__pyx_fuse_1_dist_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_1_dist_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:357:24: Exception check after calling '__pyx_fuse_1_sum' will always require the GIL to be acquired. Declare '__pyx_fuse_1_sum' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:381:35: Exception check after calling '__pyx_fuse_1_norm_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_1_norm_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      performance hint: skbio\\stats\\distance\\_cutils.pyx:389:34: Exception check after calling '__pyx_fuse_1_dist_euclidean' will always require the GIL to be acquired. Declare '__pyx_fuse_1_dist_euclidean' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:403:15: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:403:22: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:403:15: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:403:22: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:411:14: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:411:21: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:411:14: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:411:21: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:418:18: Use boundscheck(False) for faster access\n",
      "      warning: skbio\\stats\\distance\\_cutils.pyx:418:18: Use boundscheck(False) for faster access\n",
      "      [7/7] Cythonizing skbio/tree/_c_me.pyx\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-zqhsxngj\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py:270: UserWarning: Unknown distribution option: 'tests_require'\n",
      "        warnings.warn(msg)\n",
      "      running bdist_wheel\n",
      "      running build\n",
      "      running build_py\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\n",
      "      copying skbio\\test.py -> build\\lib.win-amd64-cpython-310\\skbio\n",
      "      copying skbio\\workflow.py -> build\\lib.win-amd64-cpython-310\\skbio\n",
      "      copying skbio\\_base.py -> build\\lib.win-amd64-cpython-310\\skbio\n",
      "      copying skbio\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\_indexing.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\_pairwise.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\_path.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\_repr.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\_tabular_msa.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      copying skbio\\alignment\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\n",
      "      copying skbio\\diversity\\_block.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\n",
      "      copying skbio\\diversity\\_driver.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\n",
      "      copying skbio\\diversity\\_util.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\n",
      "      copying skbio\\diversity\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\embedding\n",
      "      copying skbio\\embedding\\_embedding.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\n",
      "      copying skbio\\embedding\\_protein.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\n",
      "      copying skbio\\embedding\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\registry.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\util.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\_exception.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\_fileobject.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\_iosources.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\_warning.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      copying skbio\\io\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\base.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\io.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\missing.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_enan.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_interval.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_metadata.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_mixin.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_repr.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\_testing.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      copying skbio\\metadata\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\distance.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_alphabet.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_dna.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_genetic_code.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_grammared_sequence.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_nucleotide_mixin.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_protein.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_repr.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_rna.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_sequence.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\_substitution.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      copying skbio\\sequence\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\composition.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\gradient.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\power.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\_misc.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\_subsample.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      copying skbio\\stats\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\table\n",
      "      copying skbio\\table\\_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\table\n",
      "      copying skbio\\table\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\table\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\tests\n",
      "      copying skbio\\tests\\test_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\tests\n",
      "      copying skbio\\tests\\test_workflow.py -> build\\lib.win-amd64-cpython-310\\skbio\\tests\n",
      "      copying skbio\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_compare.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_exception.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_majority_rule.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_me.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_nj.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_tree.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_upgma.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\_utils.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      copying skbio\\tree\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_decorator.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_docstring.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_exception.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_misc.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_plotting.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_testing.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\_warning.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      copying skbio\\util\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      copying skbio\\alignment\\tests\\test_pairwise.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      copying skbio\\alignment\\tests\\test_path.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      copying skbio\\alignment\\tests\\test_ssw.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      copying skbio\\alignment\\tests\\test_tabular_msa.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      copying skbio\\alignment\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\alignment\\_lib\n",
      "      copying skbio\\alignment\\_lib\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\alignment\\_lib\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_ace.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_chao1.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_gini.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_lladser.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\_pd.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      copying skbio\\diversity\\alpha\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\n",
      "      copying skbio\\diversity\\beta\\_unifrac.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\n",
      "      copying skbio\\diversity\\beta\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\tests\n",
      "      copying skbio\\diversity\\tests\\test_block.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\tests\n",
      "      copying skbio\\diversity\\tests\\test_driver.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\tests\n",
      "      copying skbio\\diversity\\tests\\test_util.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\tests\n",
      "      copying skbio\\diversity\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_ace.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_chao1.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_gini.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_lladser.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\test_pd.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      copying skbio\\diversity\\alpha\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\n",
      "      copying skbio\\diversity\\beta\\tests\\test_unifrac.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\n",
      "      copying skbio\\diversity\\beta\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\n",
      "      copying skbio\\embedding\\tests\\test_embedding.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\n",
      "      copying skbio\\embedding\\tests\\test_protein.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\n",
      "      copying skbio\\embedding\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\binary_dm.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\biom.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\blast6.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\blast7.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\clustal.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\embed.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\embl.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\emptyfile.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\fasta.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\fastq.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\genbank.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\gff3.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\lsmat.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\newick.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\ordination.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\phylip.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\qseq.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\sample_metadata.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\stockholm.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\taxdump.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\_blast.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\_sequence_feature_vocabulary.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      copying skbio\\io\\format\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\n",
      "      copying skbio\\io\\tests\\test_iosources.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\n",
      "      copying skbio\\io\\tests\\test_registry.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\n",
      "      copying skbio\\io\\tests\\test_util.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\n",
      "      copying skbio\\io\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_binary_dm.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_biom.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_blast6.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_blast7.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_clustal.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_embed.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_embl.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_emptyfile.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_fasta.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_fastq.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_genbank.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_gff3.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_lsmat.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_newick.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_ordination.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_phylip.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_qseq.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_sample_metadata.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_sequence_feature_vocabulary.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_stockholm.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\test_taxdump.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      copying skbio\\io\\format\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_enan.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_intersection.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_interval.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_io.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_metadata.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_metadata_column.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_missing.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\test_mixin.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      copying skbio\\metadata\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_alphabet.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_distance.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_dna.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_genetic_code.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_grammared_sequence.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_nucleotide_sequences.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_protein.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_rna.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_sequence.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\test_substitution.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      copying skbio\\sequence\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\sequence\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_anosim.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_bioenv.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_mantel.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_permanova.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_permdisp.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\_utils.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      copying skbio\\stats\\distance\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\n",
      "      copying skbio\\stats\\evolve\\_hommola.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\n",
      "      copying skbio\\stats\\evolve\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_canonical_correspondence_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_correspondence_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_ordination_results.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_principal_coordinate_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_redundancy_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\_utils.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      copying skbio\\stats\\ordination\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\test_composition.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\test_gradient.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\test_misc.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\test_power.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\test_subsample.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      copying skbio\\stats\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_anosim.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_bioenv.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_mantel.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_permanova.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\test_permdisp.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      copying skbio\\stats\\distance\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\\tests\n",
      "      copying skbio\\stats\\evolve\\tests\\test_hommola.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\\tests\n",
      "      copying skbio\\stats\\evolve\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\evolve\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_canonical_correspondence_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_correspondence_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_ordination_results.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_principal_coordinate_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_redundancy_analysis.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\test_util.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      copying skbio\\stats\\ordination\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\table\\tests\n",
      "      copying skbio\\table\\tests\\test_base.py -> build\\lib.win-amd64-cpython-310\\skbio\\table\\tests\n",
      "      copying skbio\\table\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\table\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_compare.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_majority_rule.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_me.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_nj.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_tree.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\test_upgma.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      copying skbio\\tree\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\tree\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_decorator.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_docstring.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_misc.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_plotting.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_testing.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\test_warning.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      copying skbio\\util\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\skbio\\util\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\\faith-pd.txt -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\\otu-table.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\\README.md -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\\tree.nwk -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\alpha\\tests\\data\\qiime-191-tt\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\otu-table.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\README.md -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\tree.nwk -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\unweighted_unifrac_dm.txt -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\weighted_normalized_unifrac_dm.txt -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      copying skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\\weighted_unifrac_dm.txt -> build\\lib.win-amd64-cpython-310\\skbio\\diversity\\beta\\tests\\data\\qiime-191-tt\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\\data\n",
      "      copying skbio\\embedding\\tests\\data\\embed1.txt.npy -> build\\lib.win-amd64-cpython-310\\skbio\\embedding\\tests\\data\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\big5_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\big5_file.bz2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\big5_file.gz -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\example_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\example_file.bz2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\example_file.gz -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\real_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      copying skbio\\io\\tests\\data\\real_file_2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\tests\\data\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_custom_minimal -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_custom_mixed_nans -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_custom_multi_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_custom_single_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_default_multi_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_default_single_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_invalid_column_types -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_invalid_number_of_columns -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast6_invalid_type_in_column -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_custom_minimal -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_custom_mixed_nans -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_custom_multi_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_custom_single_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_default_multi_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_default_single_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_differing_fields -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_for_sniffer -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_for_sniffer_2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_gibberish -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_no_data -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_too_many_columns -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\blast7_invalid_unrecognized_field -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embed1.txt.npy -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embed2.txt.npy -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_constructed -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_feature_level_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_feature_level_record_no_FT -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_multi_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_single_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_single_record_lower -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_single_record_simple -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_single_record_upper -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_uniprot_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\embl_w_beginning_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\empty -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_diff_ids.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_double_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_double_seq.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_long_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_no_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_del.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_escape.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_null.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_space.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_tab.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_unit_sep.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_qual_vtab.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_short_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_spaces.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_tabs.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_at_plus.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_at_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_at_seq.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_in_plus.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_in_qual.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_in_seq.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\error_trunc_in_title.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_10_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_3_seqs_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_3_seqs_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_5_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_5_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_6_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_6_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_blanks_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_blank_lines_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_description_newline_replacement_empty_str -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_description_newline_replacement_multi_char -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_description_newline_replacement_none -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_id_whitespace_replacement_empty_str -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_id_whitespace_replacement_multi_char -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_id_whitespace_replacement_none -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_after_10_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_blank_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_blank_line_within_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_blank_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_legacy_format -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_missing_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_missing_seq_data_first -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_missing_seq_data_last -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_missing_seq_data_middle -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_whitespace_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_whitespace_only_line_within_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_invalid_whitespace_only_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_max_width_1 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_max_width_5 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_mixed_qual_scores -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_multi_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_multi_seq_roundtrip -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_multi_seq_spaces -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_prot_seqs_odd_labels -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_bio_seq_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_bio_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_dna_seq_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_dna_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_prot_seq_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_prot_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_rna_seq_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_rna_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_single_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_tabular_msa_different_type -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_ws_lines_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fasta_ws_lines_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_5_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_5_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_blank_lines -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_after_plus -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_after_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_in_seq_at_symbol -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_within_qual -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_blank_within_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_missing_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_missing_seq_data -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_ws_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_ws_line_after_plus -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_ws_line_after_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_ws_line_within_qual -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_invalid_ws_line_within_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_blank_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_blank_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_blank_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_seq_sanger -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_whitespace_stripping -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_ws_lines_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_ws_lines_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_multi_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_single_seq_illumina1.3 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_single_seq_illumina1.8 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_whitespace_only_lines -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_wrapping_as_illumina_no_description -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_wrapping_as_sanger_no_description -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_wrapping_original_sanger_no_description -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_writer_illumina1.3_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_writer_sanger_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\fastq_writer_sanger_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_5_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_6_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_missing_locus_name -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_multi_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_single_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_single_record_lower -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_single_record_upper -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\genbank_w_beginning_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\gff3_bad_missing_directive -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\gff3_bad_wrong_columns -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\gff3_dna -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\gff3_multi_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\gff3_single_record -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\illumina_full_range_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\illumina_full_range_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\illumina_full_range_original_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_and_blast7_default -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_invalid_differing_fields -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_invalid_too_many_columns -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_mixed_nans -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_multi_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\legacy9_single_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\longreads_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\longreads_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\longreads_original_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_dna_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_dna_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_dna_original_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_rna_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_rna_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\misc_rna_original_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error1 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error10 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error11 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error12 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error13 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error14 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error15 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error16 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error17 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error18 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error19 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error20 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error21 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error22 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error23 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error24 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error3 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error4 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error5 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error6 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error7 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error8 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_error9 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_example2_scores -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_example3_scores -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_CCA_site -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_CCA_site_constraints -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_CCA_species -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_PCoA_site -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_RDA_site -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_RDA_site_constraints -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_exp_Ordination_RDA_species -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_L&L_CA_data_scores -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\ordination_PCoA_sample_data_3_scores -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\pdb_hits.txt -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_dna_3_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_empty_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_empty_line_before_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_empty_line_between_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_header_too_long -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_header_too_short -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_no_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_seq_too_long -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_seq_too_short -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_too_few_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_too_many_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_zero_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_invalid_zero_seq_len -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_single_seq_long -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_single_seq_short -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_two_chunks -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_variable_length_ids -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_varied_whitespace_in_seqs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_whitespace_in_header_1 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_whitespace_in_header_2 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\phylip_whitespace_in_header_3 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\prot.emb -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\prot_vec.emb -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_filter -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_lane -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_read -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_tile -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_x -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_invalid_y -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_multi_seq_illumina1.3 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qseq_single_seq_sanger -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_2_seqs_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_defaults_desc_mismatch -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_defaults_extra -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_defaults_id_mismatch -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_defaults_length_mismatch -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_3_seqs_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_5_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_5_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_6_blanks_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_6_ws_lines_start_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_blanks_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_blank_lines_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_description_newline_replacement_empty_str -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_description_newline_replacement_multi_char -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_description_newline_replacement_none -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_id_whitespace_replacement_empty_str -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_id_whitespace_replacement_multi_char -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_id_whitespace_replacement_none -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_blank_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_blank_line_within_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_blank_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_legacy_format -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_missing_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_missing_qual_scores_first -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_missing_qual_scores_last -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_missing_qual_scores_middle -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_qual_scores_float -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_qual_scores_negative -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_qual_scores_over_255 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_qual_scores_string -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_whitespace_line_in_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_whitespace_only_sequence -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_invalid_ws_line_after_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_max_width_1 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_max_width_5 -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_multi_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_multi_seq_roundtrip -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_prot_seqs_odd_labels -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_bio_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_dna_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_nuc_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_prot_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_rna_seq_non_defaults -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_single_seq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_tabular_msa_different_type -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_ws_lines_between_records -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\qual_ws_lines_end_of_file -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-comments-comment-char-id.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-comments-mixed-case.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-comments.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-complete-types-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-empty-rows.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-leading-trailing-whitespace-split-id.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sample-metadata-leading-trailing-whitespace.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sanger_full_range_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sanger_full_range_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\sanger_full_range_original_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\solexa_full_range_original_solexa.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_all_data_types -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_blank_lines -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_data_only -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_different_padding -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_differing_gc_data_length -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_differing_gr_data_length -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_differing_seq_lengths -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_duplicate_gc -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_duplicate_gr -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_duplicate_sequence_names -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_duplicate_tree_ids -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_extensive -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_extensive_mixed -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_invalid_data_type -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_invalid_nonexistent_gr -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_invalid_nonexistent_gs -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_malformed_data_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_malformed_gc_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_malformed_gf_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_malformed_gr_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_malformed_gs_line -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_metadata_only -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_minimal -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_missing_footer -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_missing_header -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_missing_reference_items -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_missing_rn_tag -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multiple_msa -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multiple_multi_line_trees -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multiple_references -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multiple_trees -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multi_line_tree_no_id -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_multi_line_tree_with_id -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_nonstring_labels -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_no_data -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_rna -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_gf_no_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_gf_with_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_gs_no_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_gs_with_whitespace -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_references -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_runon_references_mixed -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_single_reference -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_single_tree_without_id -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_single_tree_with_id -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_two_of_each_metadata -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\stockholm_whitespace_only_lines -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\taxdump_names.dmp -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\taxdump_nodes.dmp -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\tiny_embedding_file.npz -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\tsv_10_fields -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\tsv_8_fields -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\whitespace_only -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\wrapping_as_illumina.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\wrapping_as_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      copying skbio\\io\\format\\tests\\data\\wrapping_original_sanger.fastq -> build\\lib.win-amd64-cpython-310\\skbio\\io\\format\\tests\\data\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\column-name-conflicts-with-id-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\comments-and-empty-rows-only.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\data-longer-than-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\directive-after-directives-section.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\directive-before-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\directive-longer-than-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\duplicate-column-names-with-whitespace.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\duplicate-column-names.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\duplicate-directives.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\duplicate-ids-with-whitespace.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\duplicate-ids.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\empty-column-name.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\empty-file -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\empty-id.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\header-only-with-comments-and-empty-rows.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\header-only.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\id-conflicts-with-id-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\invalid-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\missing-unknown-scheme.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\non-utf-8.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\qiime1-empty.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\simple-utf-16be.txt -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\simple-utf-16le.txt -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\types-directive-non-numeric.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\unrecognized-column-type.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\unrecognized-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\whitespace-only-column-name.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      copying skbio\\metadata\\tests\\data\\invalid\\whitespace-only-id.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\invalid\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\all-cells-padded.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\biom-observation-metadata.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\BOM-simple.txt -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\case-insensitive-types-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\column-order.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\comments.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\complete-types-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\complete-types-directive_w_sk.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\empty-rows.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\empty-types-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\jagged-trailing-columns.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\jagged.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\leading-trailing-whitespace.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\mac-line-endings.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\minimal.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\missing-data.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\missing-insdc-no-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\missing-insdc.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\no-columns.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\no-id-or-column-name-type-cast.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\no-newline-at-eof.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\non-standard-characters.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\numeric-column.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\override-insdc.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\partial-types-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\qiime1.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\qiita-preparation-information.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\qiita-sample-information.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\recommended-ids.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\rows-shorter-than-header.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\simple-with-directive.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\simple.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\simple.txt -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\single-column.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\single-id.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\trailing-columns.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\unix-line-endings.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      copying skbio\\metadata\\tests\\data\\valid\\windows-line-endings.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\metadata\\tests\\data\\valid\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\cr_data_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\cr_data_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\cr_no_data_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\cr_no_data_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\gr_wo_msg_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\gr_wo_msg_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\gr_w_msg_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\gr_w_msg_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\vr_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\vr_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\vr_real_out -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      copying skbio\\stats\\tests\\data\\vr_real_raw -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\tests\\data\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\bioenv_df_vegan.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\bioenv_dm_vegan.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\bioenv_exp_results_vegan.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\df.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\df_extra_column.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\dm.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\dm2.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\dm3.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\dm4.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\dm_reordered.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\exp_results.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\exp_results_different_column_order.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\exp_results_single_column.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\frameSeries_dm.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\frameSeries_grouping.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\mantel_env_dm_vegan.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\mantel_veg_dm_vegan.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\moving_pictures_dm.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\moving_pictures_mf.tsv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_all_dms.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_dm_dm2.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_duplicate_dms.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_minimal.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_minimal_with_labels.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_na_p_value.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      copying skbio\\stats\\distance\\tests\\data\\pwmantel_exp_results_reordered_distance_matrices.txt -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\distance\\tests\\data\n",
      "      creating build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_biplot_scaling1 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_biplot_scaling2 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_sample_constraints_scaling1 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_sample_constraints_scaling2 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_site_scaling1_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_site_scaling2_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_species_scaling1_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_species_scaling2_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_X -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example2_Y -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_biplot_scaling1 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_biplot_scaling2 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_sample_constraints_scaling1 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_sample_constraints_scaling2 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_site_scaling1_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_site_scaling2_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_species_scaling1_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_species_scaling2_from_vegan -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_X -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\example3_Y -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\exp_PCoAEigenResults_site -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\exp_PCoAzeros_site -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\L&L_CA_data -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_biplot_descriptors -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_biplot_projected_descriptors -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data_12dim -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data_2 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data_3 -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data_3_eigh_ref_3dim -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_sample_data_3_fsvd_ref_3dim -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\PCoA_skbio -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\varechem.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\varespec.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_biplot_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_constraints_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_eigvals_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_propexpl_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_sites_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      copying skbio\\stats\\ordination\\tests\\data\\vare_rda_species_from_vegan.csv -> build\\lib.win-amd64-cpython-310\\skbio\\stats\\ordination\\tests\\data\n",
      "      running build_ext\n",
      "      building 'skbio.metadata._intersection' extension\n",
      "      error: Microsoft Visual C++ 14.0 or greater is required. Get it with \"Microsoft C++ Build Tools\": https://visualstudio.microsoft.com/visual-cpp-build-tools/\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for scikit-bio\n",
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for biom-format (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [332 lines of output]\n",
      "      <string>:15: SetuptoolsDeprecationWarning: The test command is disabled and references to it are deprecated.\n",
      "      !!\n",
      "      \n",
      "              ********************************************************************************\n",
      "              Please remove any references to `setuptools.command.test` in all supported versions of the affected package.\n",
      "      \n",
      "              This deprecation is overdue, please update your project and remove deprecated\n",
      "              calls to avoid build errors in the future.\n",
      "              ********************************************************************************\n",
      "      \n",
      "      !!\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py:270: UserWarning: Unknown distribution option: 'tests_require'\n",
      "        warnings.warn(msg)\n",
      "      running bdist_wheel\n",
      "      running build\n",
      "      running build_py\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\err.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\exception.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\parse.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\table.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\util.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\__init__.py -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\installation_informer.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\metadata_adder.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\metadata_exporter.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_converter.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_head.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_ids.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_normalizer.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_subsetter.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_summarizer.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\table_validator.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\uc_processor.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\util.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      copying biom\\cli\\__init__.py -> build\\lib.win-amd64-cpython-310\\biom\\cli\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\long_lines.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\test_err.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\test_parse.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\test_table.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\test_util.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      copying biom\\tests\\__init__.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_add_metadata.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_show_install_info.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_subset_table.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_summarize_table.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_table_converter.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_table_normalizer.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_uc_processor.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\test_validate_table.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      copying biom\\tests\\test_cli\\__init__.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\__init__.py -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      running egg_info\n",
      "      writing biom_format.egg-info\\PKG-INFO\n",
      "      writing dependency_links to biom_format.egg-info\\dependency_links.txt\n",
      "      writing entry points to biom_format.egg-info\\entry_points.txt\n",
      "      writing requirements to biom_format.egg-info\\requires.txt\n",
      "      writing top-level names to biom_format.egg-info\\top_level.txt\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayscalars.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarraytypes.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ufuncobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayscalars.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarraytypes.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ufuncobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\arrayscalars.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarrayobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ndarraytypes.h won't be automatically included in the manifest: the path must be relative\n",
      "      dependency C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\numpy\\_core\\include\\numpy\\ufuncobject.h won't be automatically included in the manifest: the path must be relative\n",
      "      reading manifest file 'biom_format.egg-info\\SOURCES.txt'\n",
      "      reading manifest template 'MANIFEST.in'\n",
      "      warning: no directories found matching 'support_files'\n",
      "      warning: no directories found matching 'tests\\test_data'\n",
      "      no previously-included directories found matching 'docs\\_build'\n",
      "      warning: no previously-included files matching '*.pyc' found anywhere in distribution\n",
      "      warning: no previously-included files matching '*.pyo' found anywhere in distribution\n",
      "      warning: no previously-included files matching '.git' found anywhere in distribution\n",
      "      warning: no previously-included files matching '*.so' found anywhere in distribution\n",
      "      warning: no previously-included files matching '.*.swp' found anywhere in distribution\n",
      "      adding license file 'COPYING.txt'\n",
      "      writing manifest file 'biom_format.egg-info\\SOURCES.txt'\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\setuptools\\command\\build_py.py:212: _Warning: Package 'biom.assets' is absent from the `packages` configuration.\n",
      "      !!\n",
      "      \n",
      "              ********************************************************************************\n",
      "              ############################\n",
      "              # Package would be ignored #\n",
      "              ############################\n",
      "              Python recognizes 'biom.assets' as an importable package[^1],\n",
      "              but it is absent from setuptools' `packages` configuration.\n",
      "      \n",
      "              This leads to an ambiguous overall configuration. If you want to distribute this\n",
      "              package, please make sure that 'biom.assets' is explicitly added\n",
      "              to the `packages` configuration field.\n",
      "      \n",
      "              Alternatively, you can also rely on setuptools' discovery methods\n",
      "              (for example by using `find_namespace_packages(...)`/`find_namespace:`\n",
      "              instead of `find_packages(...)`/`find:`).\n",
      "      \n",
      "              You can read more about \"package discovery\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html\n",
      "      \n",
      "              If you don't want 'biom.assets' to be distributed and are\n",
      "              already explicitly excluding 'biom.assets' via\n",
      "              `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,\n",
      "              you can try to use `exclude_package_data`, or `include-package-data=False` in\n",
      "              combination with a more fine grained `package-data` configuration.\n",
      "      \n",
      "              You can read more about \"package data files\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/datafiles.html\n",
      "      \n",
      "      \n",
      "              [^1]: For Python, any directory (with suitable naming) can be imported,\n",
      "                    even if it does not contain any `.py` files.\n",
      "                    On the other hand, currently there is no concept of package data\n",
      "                    directory, all directories are treated like packages.\n",
      "              ********************************************************************************\n",
      "      \n",
      "      !!\n",
      "        check.warn(importable)\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\setuptools\\command\\build_py.py:212: _Warning: Package 'biom.tests.bench_tables' is absent from the `packages` configuration.\n",
      "      !!\n",
      "      \n",
      "              ********************************************************************************\n",
      "              ############################\n",
      "              # Package would be ignored #\n",
      "              ############################\n",
      "              Python recognizes 'biom.tests.bench_tables' as an importable package[^1],\n",
      "              but it is absent from setuptools' `packages` configuration.\n",
      "      \n",
      "              This leads to an ambiguous overall configuration. If you want to distribute this\n",
      "              package, please make sure that 'biom.tests.bench_tables' is explicitly added\n",
      "              to the `packages` configuration field.\n",
      "      \n",
      "              Alternatively, you can also rely on setuptools' discovery methods\n",
      "              (for example by using `find_namespace_packages(...)`/`find_namespace:`\n",
      "              instead of `find_packages(...)`/`find:`).\n",
      "      \n",
      "              You can read more about \"package discovery\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html\n",
      "      \n",
      "              If you don't want 'biom.tests.bench_tables' to be distributed and are\n",
      "              already explicitly excluding 'biom.tests.bench_tables' via\n",
      "              `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,\n",
      "              you can try to use `exclude_package_data`, or `include-package-data=False` in\n",
      "              combination with a more fine grained `package-data` configuration.\n",
      "      \n",
      "              You can read more about \"package data files\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/datafiles.html\n",
      "      \n",
      "      \n",
      "              [^1]: For Python, any directory (with suitable naming) can be imported,\n",
      "                    even if it does not contain any `.py` files.\n",
      "                    On the other hand, currently there is no concept of package data\n",
      "                    directory, all directories are treated like packages.\n",
      "              ********************************************************************************\n",
      "      \n",
      "      !!\n",
      "        check.warn(importable)\n",
      "      C:\\Users\\histe\\AppData\\Local\\Temp\\pip-build-env-ahjpuk4v\\overlay\\Lib\\site-packages\\setuptools\\command\\build_py.py:212: _Warning: Package 'biom.tests.test_cli.test_data' is absent from the `packages` configuration.\n",
      "      !!\n",
      "      \n",
      "              ********************************************************************************\n",
      "              ############################\n",
      "              # Package would be ignored #\n",
      "              ############################\n",
      "              Python recognizes 'biom.tests.test_cli.test_data' as an importable package[^1],\n",
      "              but it is absent from setuptools' `packages` configuration.\n",
      "      \n",
      "              This leads to an ambiguous overall configuration. If you want to distribute this\n",
      "              package, please make sure that 'biom.tests.test_cli.test_data' is explicitly added\n",
      "              to the `packages` configuration field.\n",
      "      \n",
      "              Alternatively, you can also rely on setuptools' discovery methods\n",
      "              (for example by using `find_namespace_packages(...)`/`find_namespace:`\n",
      "              instead of `find_packages(...)`/`find:`).\n",
      "      \n",
      "              You can read more about \"package discovery\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html\n",
      "      \n",
      "              If you don't want 'biom.tests.test_cli.test_data' to be distributed and are\n",
      "              already explicitly excluding 'biom.tests.test_cli.test_data' via\n",
      "              `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,\n",
      "              you can try to use `exclude_package_data`, or `include-package-data=False` in\n",
      "              combination with a more fine grained `package-data` configuration.\n",
      "      \n",
      "              You can read more about \"package data files\" on setuptools documentation page:\n",
      "      \n",
      "              - https://setuptools.pypa.io/en/latest/userguide/datafiles.html\n",
      "      \n",
      "      \n",
      "              [^1]: For Python, any directory (with suitable naming) can be imported,\n",
      "                    even if it does not contain any `.py` files.\n",
      "                    On the other hand, currently there is no concept of package data\n",
      "                    directory, all directories are treated like packages.\n",
      "              ********************************************************************************\n",
      "      \n",
      "      !!\n",
      "        check.warn(importable)\n",
      "      copying biom\\_filter.pyx -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\_subsample.pyx -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      copying biom\\_transform.pyx -> build\\lib.win-amd64-cpython-310\\biom\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\assets\n",
      "      copying biom\\assets\\exercise_api.py -> build\\lib.win-amd64-cpython-310\\biom\\assets\n",
      "      copying biom\\assets\\exercise_cli.sh -> build\\lib.win-amd64-cpython-310\\biom\\assets\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x1000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x100x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10000x10x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x1000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x100x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\1000x10x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x1000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x100x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\100x10x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x1000x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x100x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.001_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.001_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.010_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.010_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.100_bench.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      copying biom\\tests\\bench_tables\\10x10x0.100_bench_hdf5.biom.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\bench_tables\n",
      "      creating build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\\test_data\n",
      "      copying biom\\tests\\test_cli\\test_data\\json_obs_collapsed.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\\test_data\n",
      "      copying biom\\tests\\test_cli\\test_data\\json_sample_collapsed.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\\test_data\n",
      "      copying biom\\tests\\test_cli\\test_data\\test.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\\test_data\n",
      "      copying biom\\tests\\test_cli\\test_data\\test.json -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_cli\\test_data\n",
      "      copying biom\\tests\\test_data\\bad_table.txt -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\bad_table.txt.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\edgecase_issue_952.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\empty.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\no-contents.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\test.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\test.json -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\test.json.gz -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      copying biom\\tests\\test_data\\test_grp_metadata.biom -> build\\lib.win-amd64-cpython-310\\biom\\tests\\test_data\n",
      "      running build_ext\n",
      "      building 'biom._filter' extension\n",
      "      error: Microsoft Visual C++ 14.0 or greater is required. Get it with \"Microsoft C++ Build Tools\": https://visualstudio.microsoft.com/visual-cpp-build-tools/\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for biom-format\n",
      "ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (scikit-bio, biom-format)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scikit-bio\n",
      "  Using cached scikit-bio-0.6.3.tar.gz (5.4 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: still running...\n",
      "  Installing build dependencies: still running...\n",
      "  Installing build dependencies: still running...\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: requests>=2.20.0 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (2.32.3)\n",
      "Requirement already satisfied: decorator>=3.4.2 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (5.1.1)\n",
      "Requirement already satisfied: natsort>=4.0.3 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (8.4.0)\n",
      "Requirement already satisfied: numpy>=1.17.0 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (1.26.4)\n",
      "Requirement already satisfied: pandas>=1.5.0 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (2.2.3)\n",
      "Requirement already satisfied: scipy>=1.9.0 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from scikit-bio) (1.15.1)\n",
      "Collecting h5py>=3.6.0 (from scikit-bio)\n",
      "  Using cached h5py-3.12.1-cp310-cp310-win_amd64.whl.metadata (2.5 kB)\n",
      "Collecting biom-format>=2.1.16 (from scikit-bio)\n",
      "  Using cached biom-format-2.1.16.tar.gz (11.7 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Collecting statsmodels>=0.14.0 (from scikit-bio)\n",
      "  Using cached statsmodels-0.14.4-cp310-cp310-win_amd64.whl.metadata (9.5 kB)\n",
      "Collecting patsy>=0.5.0 (from scikit-bio)\n",
      "  Using cached patsy-1.0.1-py2.py3-none-any.whl.metadata (3.3 kB)\n",
      "Collecting click (from biom-format>=2.1.16->scikit-bio)\n",
      "  Using cached click-8.1.8-py3-none-any.whl.metadata (2.3 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from pandas>=1.5.0->scikit-bio) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from pandas>=1.5.0->scikit-bio) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from pandas>=1.5.0->scikit-bio) (2023.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from requests>=2.20.0->scikit-bio) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from requests>=2.20.0->scikit-bio) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from requests>=2.20.0->scikit-bio) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from requests>=2.20.0->scikit-bio) (2024.12.14)\n",
      "Requirement already satisfied: packaging>=21.3 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from statsmodels>=0.14.0->scikit-bio) (24.2)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from python-dateutil>=2.8.2->pandas>=1.5.0->scikit-bio) (1.16.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\histe\\anaconda3\\envs\\biom480a5\\lib\\site-packages (from click->biom-format>=2.1.16->scikit-bio) (0.4.6)\n",
      "Using cached h5py-3.12.1-cp310-cp310-win_amd64.whl (3.0 MB)\n",
      "Using cached patsy-1.0.1-py2.py3-none-any.whl (232 kB)\n",
      "Using cached statsmodels-0.14.4-cp310-cp310-win_amd64.whl (9.8 MB)\n",
      "Using cached click-8.1.8-py3-none-any.whl (98 kB)\n",
      "Building wheels for collected packages: scikit-bio, biom-format\n",
      "  Building wheel for scikit-bio (pyproject.toml): started\n",
      "  Building wheel for scikit-bio (pyproject.toml): finished with status 'error'\n",
      "  Building wheel for biom-format (pyproject.toml): started\n",
      "  Building wheel for biom-format (pyproject.toml): finished with status 'error'\n",
      "Failed to build scikit-bio biom-format\n"
     ]
    }
   ],
   "source": [
    "# Let's try to install the package scikit-bio.  This package is not installed by default in the\n",
    "# python environment, so we will need to install it.  We can use the pip command to install the\n",
    "# package.  The command is \"pip install\" followed by the name of the package.  The exclamation\n",
    "# point is used to indicate that the command is a shell command.  \n",
    "# https://scikit.bio\n",
    "\n",
    "!pip install scikit-bio\n",
    "import skbio\n",
    "print('')\n",
    "print('The version of scikit-bio installed:')\n",
    "print(skbio.__version__)\n",
    "\n",
    "# Note that this could take several minutes.  We are not going to use the package in this notebook,\n",
    "# so you can skip this cell if it takes too long. To interrupt the installation, you can click on the\n",
    "# stop button in the toolbar above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PafXp9PlKfXO"
   },
   "source": [
    "### **1.D) os - \"operating system\"**\n",
    "\n",
    "```os```  is the main package that lets you navigate and change directories for various purposes, similar to how you move around in a command prompt or terminal. ```os``` also performs miscellaneous operating system functions such as changing directories or checking process ids.\n",
    "\n",
    "https://docs.python.org/3/library/os.html\n",
    "\n",
    "\n",
    "Another popular file path manager is ```pathlib```.\n",
    "\n",
    "https://docs.python.org/3/library/pathlib.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "QdBoojYxKQCp",
    "outputId": "29d7b573-9b8c-497d-a4fe-7c38b8d79eaf",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\histe\\OneDrive\\Documents\\GitHub\\480A5CourseMaterials\\Module1-PreliminaryPython\n"
     ]
    }
   ],
   "source": [
    "# Let's use the os module to print the current working directory\n",
    "import os\n",
    "print(os.getcwd())  #what is our current working directory (cwd)?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cTqlun-uliFG"
   },
   "source": [
    "Once again, for us to import ```os```, the ```os``` module must exist in our system path so Python can \"see\" it.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "7iP1fyEAlsay",
    "outputId": "a26ee17f-04e4-4ec4-889d-f3949d51e3d1",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We are using the following file as the package os:\n",
      "c:\\Users\\histe\\anaconda3\\envs\\BIOM480A5\\lib\\os.py\n",
      "\n",
      "Here is our current sys.path:\n",
      "['c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\python310.zip', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\DLLs', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5', '', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\win32', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\win32\\\\lib', 'c:\\\\Users\\\\histe\\\\anaconda3\\\\envs\\\\BIOM480A5\\\\lib\\\\site-packages\\\\Pythonwin', 'c:\\\\Users\\\\histe\\\\OneDrive\\\\Documents\\\\GitHub\\\\480A5CourseMaterials\\\\BIOM480A5', 'c:\\\\Users\\\\histe\\\\OneDrive\\\\Documents\\\\GitHub\\\\480A5CourseMaterials\\\\BIOM480A5']\n"
     ]
    }
   ],
   "source": [
    "print('We are using the following file as the package os:')\n",
    "print(os.__file__) # Print the actual file that we imported the os packages from\n",
    "print('')\n",
    "print('Here is our current sys.path:')\n",
    "print(sys.path)\n",
    "# Note how this file we just imported from is inside our path!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at our current directory and its files using the ```os``` module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3-f8NhJbLH1e",
    "outputId": "fb03cf7c-bda6-4c18-a929-e5ae232e5a9d",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#list all files in a directory, '.' denotes our current working directory\n",
    "os.listdir('.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "H_s-6Nz_XrhJ"
   },
   "source": [
    "```os``` uses Unix-like paths:\n",
    "\n",
    "``` .``` a single period denotes my current working directory\n",
    "\n",
    "``` ..``` a single period denotes one parent folder above my current working directory\n",
    "\n",
    "``` ~``` a tilde represents the current users home directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "za-SJ9P6YbxV",
    "outputId": "58144aaf-73ff-416f-c7d7-9186f931c2bb",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files and folders in my current directory\n",
      "['cheatsheets-1.png', 'inclassReview_M1Aw.ipynb', 'iris.png', 'M1A_Intro_to_Python.ipynb', 'M1B_Packages_and_File_Management.ipynb', 'M1C_Python_Loops_and_Functions.ipynb', 'M1D_NumPy_and_Linear_Algebra.ipynb', 'M1E_Eigen_Values_Vectors_Decompositions_and_Transformations.ipynb', 'M1E_InClassExample_PCA.ipynb', 'M1F_Matplotlib.ipynb', 'sampleData', 'test.bin', 'test.txt', 'test_csv.csv']\n",
      "\n",
      "Files and folders one directory above my current directory\n",
      "['.git', '.gitignore', 'BIOM480A5', 'check_environment.py', 'Module1-PreliminaryPython', 'Module2-Bioinformatics', 'muscle.exe', 'README.md']\n"
     ]
    }
   ],
   "source": [
    "print('Files and folders in my current directory')\n",
    "print(os.listdir('.'))\n",
    "print('')\n",
    "print('Files and folders one directory above my current directory')\n",
    "print(os.listdir('..'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "QBcoaAM2Yw6N",
    "outputId": "f19c6ea1-ac40-4d80-9104-61de3a4fd8f3",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C:\\Users\\histe\n"
     ]
    }
   ],
   "source": [
    "# Use os to show the current user's home directory\n",
    "print(os.path.expanduser('~'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Quiz to practice importing packages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set path and load quiz functions.\n",
    "sys.path.append(os.path.abspath('../BIOM480A5'))\n",
    "from BIOM480A5_Quizzes import multiple_choice_quiz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiple_choice_quiz('importing_and_using_modules')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now might also be a good time to practice your **command line functions** using the following quiz."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# refresh import of BIOM480A5_Quizzes to get the updated version\n",
    "importlib.reload(BIOM480A5_Quizzes)\n",
    "from BIOM480A5_Quizzes import multiple_choice_quiz\n",
    "multiple_choice_quiz('command_line_tools')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XntLwYNAGan2"
   },
   "source": [
    "## **2) Reading and Writing files**\n",
    "\n",
    "Opening a file is done with \"open\" and opens a buffer to a particular file on your hard drive. It's best practice to use ``` with open() as object:``` so the buffer closes on its own after the execution of code within the ``` with ``` is over. Typically you may see people use ```f``` to denote their file buffer object.\n",
    "\n",
    " ```python\n",
    "with open('example.txt', 'r') as f:\n",
    "    print(f.readlines)\n",
    "    \n",
    " ```\n",
    "\n",
    " ```open``` takes an additional flag to denote whether you are reading, ```'r'```, writing, ```'w'```, or appending ```'a'``` a file.\n",
    "\n",
    "Additonally, if you are not reading a text file, you may have to read / write a file in binary mode, ```'b'```, by appending b to your flag (```'wb'``` would mean write in binary mode)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KKXaUdW_Gkbd",
    "outputId": "8e8d90c7-fe19-4184-cab9-fd28ba9fde41",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"hi i'm some lines to write to a file \\n\", '1kjjkl\\n', '2\\n', '3\\n']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_lines = [\"hi i'm some lines to write to a file \\n\", \"1kjjkl\\n\",\"2\\n\",\"3\\n\"]\n",
    "# \\n is a newline character, i.e., pressing enter\n",
    "\n",
    "# im writing a regular text file, 'w'\n",
    "with open('test.txt','w') as f:\n",
    "  f.writelines(text_lines)\n",
    "\n",
    "# im opening a file to read it, 'r'\n",
    "with open('test.txt','r') as f:\n",
    "  read_in_lines = f.readlines()\n",
    "\n",
    "read_in_lines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ANovnKhdPwnA"
   },
   "source": [
    "Try opening the file we just made.  In VS Code, you can see this file on the right, just under the folder icon!\n",
    "\n",
    "We can also write in binary mode and see what that looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PI5MVzT4Hq6E",
    "outputId": "8c452080-d12a-46f7-9196-2c7f53f44a95",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing array to bin file:\n",
      "[5 1 4 2 3 6]\n",
      "\n",
      "here is what was stored in file:\n",
      "\n",
      "\n",
      "array after read back in:\n",
      "[4294967301, 8589934596, 25769803779]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'hexdump' is not recognized as an internal or external command,\n",
      "operable program or batch file.\n"
     ]
    }
   ],
   "source": [
    "# import the numpy package\n",
    "import numpy as np\n",
    "\n",
    "# lets write some arbitrary array to a file as a binary file\n",
    "print('writing array to bin file:')\n",
    "array = np.array([5,1,4,2,3,6])\n",
    "print(array)\n",
    "\n",
    "# open a file called 'test.bin' in write binary mode\n",
    "with open('test.bin','wb') as f:\n",
    "  for i in array:\n",
    "    f.write(i)\n",
    "\n",
    "# lets use the kernel to print out what is stored in the binary file\n",
    "print('')\n",
    "print('here is what was stored in file:')\n",
    "!hexdump -C test.bin\n",
    "print('')\n",
    "\n",
    "# lets read that back into python\n",
    "with open('test.bin','rb') as f:\n",
    "  bitarray = []\n",
    "  byte = f.read(1)\n",
    "  bitarray.append(byte)\n",
    "  while byte != b\"\":\n",
    "    byte = f.read(1)\n",
    "    bitarray.append(byte)\n",
    "\n",
    "# now let's convert back from a byte array to a usable numpy array\n",
    "print('')\n",
    "read_array = []\n",
    "for i in range(int(len(bitarray)/8) ):\n",
    "  read_array.append(b''.join(bitarray[i*8:8 + i*8]))\n",
    "read_array = [int.from_bytes(x,byteorder='little') for x in read_array]\n",
    "\n",
    "print('array after read back in:')\n",
    "print(read_array)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "gGUrc4qghP3E"
   },
   "source": [
    "We will see more about numpy later, but for now you can also save a numpy array to its own file format using ```np.save(file, array)```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "LUrrhOXShNuw",
    "outputId": "9082a5b3-a0c7-4cd6-9895-ee97d111000c",
    "tags": []
   },
   "outputs": [],
   "source": [
    "np.save('test.npy', array)\n",
    "# print out this new file\n",
    "!head test.npy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Quiz to practice File Operations**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Which command would you use to perform the following operation on the file X:\n",
      " print the contents of the file X to the screen ?\n",
      "0: X.readline()\n",
      "1: print(X.read())\n",
      "2: with open(X) as f1, open(Y) as f2: with open(Z, \"w\") as f3: f3.write(f1.read() + f2.read())\n",
      "3: X.seekable()\n",
      "4: X.writable()\n",
      "5: open(X, \"a\")\n",
      "You answered 0 questions correctly. Goodbye!\n"
     ]
    }
   ],
   "source": [
    "multiple_choice_quiz('file_operations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "K84SSQMcGj22"
   },
   "source": [
    "## **3) Pandas file management**\n",
    "\n",
    "```pandas``` is a library that extends NumPy by adding row and column headers. The primary data format it uses is called a \"DataFrame.\" Pandas will read a plethora of file formats into the dataframe format for easy object oriented manipulation within python. When you are done with the manipulation, you can write the file back out or return the underlying NumPy array. We wont go into great detail with ```pandas``` in this course, however it is a very powerful tool for manipulating data just as ```numpy``` is for numerical analyses.\n",
    "\n",
    "https://pandas.pydata.org/docs/getting_started/index.html#getting-started\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jzphOCQVM3AT"
   },
   "source": [
    "### **3.A) Opening a file with Pandas**\n",
    "\n",
    "Let's open some example data into Jupyter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "id": "ilUoElr2Lk8K",
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#Let's read some pre-generated data that was included for you in the CourseMaterials repository.\n",
    "housing_dataframe = pd.read_csv('C:/Users/histe/OneDrive/Documents/GitHub/480A5CourseMaterials/Module1-PreliminaryPython/sampleData/california_housing_test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 487
    },
    "id": "jq-G_oXf5EgI",
    "outputId": "7d79cd46-5d9b-4389-83e0-5b84df195d3a",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>housing_median_age</th>\n",
       "      <th>total_rooms</th>\n",
       "      <th>total_bedrooms</th>\n",
       "      <th>population</th>\n",
       "      <th>households</th>\n",
       "      <th>median_income</th>\n",
       "      <th>median_house_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-122.05</td>\n",
       "      <td>37.37</td>\n",
       "      <td>27.0</td>\n",
       "      <td>3885.0</td>\n",
       "      <td>661.0</td>\n",
       "      <td>1537.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>6.6085</td>\n",
       "      <td>344700.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-118.30</td>\n",
       "      <td>34.26</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1510.0</td>\n",
       "      <td>310.0</td>\n",
       "      <td>809.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>3.5990</td>\n",
       "      <td>176500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-117.81</td>\n",
       "      <td>33.78</td>\n",
       "      <td>27.0</td>\n",
       "      <td>3589.0</td>\n",
       "      <td>507.0</td>\n",
       "      <td>1484.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>5.7934</td>\n",
       "      <td>270500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-118.36</td>\n",
       "      <td>33.82</td>\n",
       "      <td>28.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.1359</td>\n",
       "      <td>330000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-119.67</td>\n",
       "      <td>36.33</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1241.0</td>\n",
       "      <td>244.0</td>\n",
       "      <td>850.0</td>\n",
       "      <td>237.0</td>\n",
       "      <td>2.9375</td>\n",
       "      <td>81700.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2995</th>\n",
       "      <td>-119.86</td>\n",
       "      <td>34.42</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>642.0</td>\n",
       "      <td>1258.0</td>\n",
       "      <td>607.0</td>\n",
       "      <td>1.1790</td>\n",
       "      <td>225000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996</th>\n",
       "      <td>-118.14</td>\n",
       "      <td>34.06</td>\n",
       "      <td>27.0</td>\n",
       "      <td>5257.0</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>3496.0</td>\n",
       "      <td>1036.0</td>\n",
       "      <td>3.3906</td>\n",
       "      <td>237200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2997</th>\n",
       "      <td>-119.70</td>\n",
       "      <td>36.30</td>\n",
       "      <td>10.0</td>\n",
       "      <td>956.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>693.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>2.2895</td>\n",
       "      <td>62000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2998</th>\n",
       "      <td>-117.12</td>\n",
       "      <td>34.10</td>\n",
       "      <td>40.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>3.2708</td>\n",
       "      <td>162500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999</th>\n",
       "      <td>-119.63</td>\n",
       "      <td>34.42</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1765.0</td>\n",
       "      <td>263.0</td>\n",
       "      <td>753.0</td>\n",
       "      <td>260.0</td>\n",
       "      <td>8.5608</td>\n",
       "      <td>500001.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \\\n",
       "0       -122.05     37.37                27.0       3885.0           661.0   \n",
       "1       -118.30     34.26                43.0       1510.0           310.0   \n",
       "2       -117.81     33.78                27.0       3589.0           507.0   \n",
       "3       -118.36     33.82                28.0         67.0            15.0   \n",
       "4       -119.67     36.33                19.0       1241.0           244.0   \n",
       "...         ...       ...                 ...          ...             ...   \n",
       "2995    -119.86     34.42                23.0       1450.0           642.0   \n",
       "2996    -118.14     34.06                27.0       5257.0          1082.0   \n",
       "2997    -119.70     36.30                10.0        956.0           201.0   \n",
       "2998    -117.12     34.10                40.0         96.0            14.0   \n",
       "2999    -119.63     34.42                42.0       1765.0           263.0   \n",
       "\n",
       "      population  households  median_income  median_house_value  \n",
       "0         1537.0       606.0         6.6085            344700.0  \n",
       "1          809.0       277.0         3.5990            176500.0  \n",
       "2         1484.0       495.0         5.7934            270500.0  \n",
       "3           49.0        11.0         6.1359            330000.0  \n",
       "4          850.0       237.0         2.9375             81700.0  \n",
       "...          ...         ...            ...                 ...  \n",
       "2995      1258.0       607.0         1.1790            225000.0  \n",
       "2996      3496.0      1036.0         3.3906            237200.0  \n",
       "2997       693.0       220.0         2.2895             62000.0  \n",
       "2998        46.0        14.0         3.2708            162500.0  \n",
       "2999       753.0       260.0         8.5608            500001.0  \n",
       "\n",
       "[3000 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If we just execute the variable housing_dataframe, we will see the first and last few rows of the dataframe.\n",
    "housing_dataframe "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "S1yKWhMaM-Y7"
   },
   "source": [
    "Getting columns and Logicals\n",
    "\n",
    "The (dis)advantage of pandas is that the slicing that you would do with numpy can be done with strings instead of indexes.\n",
    "\n",
    "Pros:\n",
    "\n",
    "* Comprehension\n",
    "\n",
    "It's clear which column you are using at a time since you can use text label headers.\n",
    "\n",
    "Cons:\n",
    "* Readability\n",
    "\n",
    "When stringing together long logicals slicing dataframes, the readability can quickly become incomprehensible.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 487
    },
    "id": "tyYS_GY95_z7",
    "outputId": "548ccb22-f7b7-4959-a74a-293bc6927534",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>housing_median_age</th>\n",
       "      <th>total_rooms</th>\n",
       "      <th>total_bedrooms</th>\n",
       "      <th>population</th>\n",
       "      <th>households</th>\n",
       "      <th>median_income</th>\n",
       "      <th>median_house_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>-122.42</td>\n",
       "      <td>37.76</td>\n",
       "      <td>52.0</td>\n",
       "      <td>3587.0</td>\n",
       "      <td>1030.0</td>\n",
       "      <td>2259.0</td>\n",
       "      <td>979.0</td>\n",
       "      <td>2.5403</td>\n",
       "      <td>250000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>-122.53</td>\n",
       "      <td>37.97</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1560.0</td>\n",
       "      <td>451.0</td>\n",
       "      <td>700.0</td>\n",
       "      <td>419.0</td>\n",
       "      <td>2.5125</td>\n",
       "      <td>270800.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>-121.92</td>\n",
       "      <td>37.33</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2125.0</td>\n",
       "      <td>382.0</td>\n",
       "      <td>930.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>5.2831</td>\n",
       "      <td>299500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>-118.20</td>\n",
       "      <td>33.77</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1375.0</td>\n",
       "      <td>457.0</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>2.2344</td>\n",
       "      <td>200000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>-122.43</td>\n",
       "      <td>37.74</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1514.0</td>\n",
       "      <td>314.0</td>\n",
       "      <td>724.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>5.3292</td>\n",
       "      <td>300900.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2950</th>\n",
       "      <td>-122.42</td>\n",
       "      <td>37.79</td>\n",
       "      <td>52.0</td>\n",
       "      <td>3364.0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>2112.0</td>\n",
       "      <td>1045.0</td>\n",
       "      <td>2.1343</td>\n",
       "      <td>400000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2965</th>\n",
       "      <td>-122.44</td>\n",
       "      <td>37.71</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2711.0</td>\n",
       "      <td>591.0</td>\n",
       "      <td>1848.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>3.9567</td>\n",
       "      <td>251500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2970</th>\n",
       "      <td>-122.47</td>\n",
       "      <td>37.77</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2241.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>1042.0</td>\n",
       "      <td>377.0</td>\n",
       "      <td>4.1635</td>\n",
       "      <td>398400.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2972</th>\n",
       "      <td>-118.14</td>\n",
       "      <td>34.17</td>\n",
       "      <td>52.0</td>\n",
       "      <td>2667.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>1681.0</td>\n",
       "      <td>504.0</td>\n",
       "      <td>4.0524</td>\n",
       "      <td>173100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2992</th>\n",
       "      <td>-122.33</td>\n",
       "      <td>37.39</td>\n",
       "      <td>52.0</td>\n",
       "      <td>573.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>6.2263</td>\n",
       "      <td>500001.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \\\n",
       "26      -122.42     37.76                52.0       3587.0          1030.0   \n",
       "36      -122.53     37.97                52.0       1560.0           451.0   \n",
       "70      -121.92     37.33                52.0       2125.0           382.0   \n",
       "99      -118.20     33.77                52.0       1375.0           457.0   \n",
       "109     -122.43     37.74                52.0       1514.0           314.0   \n",
       "...         ...       ...                 ...          ...             ...   \n",
       "2950    -122.42     37.79                52.0       3364.0          1100.0   \n",
       "2965    -122.44     37.71                52.0       2711.0           591.0   \n",
       "2970    -122.47     37.77                52.0       2241.0           443.0   \n",
       "2972    -118.14     34.17                52.0       2667.0           486.0   \n",
       "2992    -122.33     37.39                52.0        573.0           102.0   \n",
       "\n",
       "      population  households  median_income  median_house_value  \n",
       "26        2259.0       979.0         2.5403            250000.0  \n",
       "36         700.0       419.0         2.5125            270800.0  \n",
       "70         930.0       387.0         5.2831            299500.0  \n",
       "99        1089.0       317.0         2.2344            200000.0  \n",
       "109        724.0       301.0         5.3292            300900.0  \n",
       "...          ...         ...            ...                 ...  \n",
       "2950      2112.0      1045.0         2.1343            400000.0  \n",
       "2965      1848.0       524.0         3.9567            251500.0  \n",
       "2970      1042.0       377.0         4.1635            398400.0  \n",
       "2972      1681.0       504.0         4.0524            173100.0  \n",
       "2992       232.0        92.0         6.2263            500001.0  \n",
       "\n",
       "[184 rows x 9 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lets get only the housing data from those with an household age of > 50\n",
    "housing_dataframe[housing_dataframe['housing_median_age'  ] > 50]\n",
    "# This shows the first and last few rows of the dataframe that meet the condition. Note that the\n",
    "# age of the house is greater than 50 for all of the rows that are displayed.  Also note that the\n",
    "# index of the rows is not continuous.  This is because the rows that do not meet the condition are\n",
    "# not displayed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "9zKWs4pFLyZ1",
    "outputId": "c6bce7e3-4de4-45de-f39b-5fbfa6b9bece",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       27.0\n",
       "1       43.0\n",
       "2       27.0\n",
       "3       28.0\n",
       "4       19.0\n",
       "        ... \n",
       "2995    23.0\n",
       "2996    27.0\n",
       "2997    10.0\n",
       "2998    40.0\n",
       "2999    42.0\n",
       "Name: housing_median_age, Length: 3000, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Suppose that we only want to look at a singel column, such as the housing median age\n",
    "all_median_ages = housing_dataframe['housing_median_age']\n",
    "all_median_ages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the lenght of the original dataframe is 3000\n",
      "the lenght of the dataframe with ages above 25 is 1772\n"
     ]
    }
   ],
   "source": [
    "# Or if we only want to look at the rows that have a housing median age greater than 25\n",
    "ages_above_25 = housing_dataframe[housing_dataframe['housing_median_age'] > 25]\n",
    "ages_above_25\n",
    "\n",
    "print(f'the lenght of the original dataframe is {len(housing_dataframe)}')\n",
    "print(f'the lenght of the dataframe with ages above 25 is {len(ages_above_25)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 296
    },
    "id": "l-S55RtEL3x9",
    "outputId": "bd7162eb-3fe9-4047-a52a-23dada632274",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'probability')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkAAAAGwCAYAAABB4NqyAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAALm5JREFUeJzt3XtUlXW+x/HPltv2pKBBcTFE0ClpyMxNFzDGOhWOpl3GTkwXa6XWIi2VPZ3jhcq0Gqw8xDFFjoo1zrTU1mC3FSehUrxAUyCYU4zT5AVLGA+6AsdOIPCcP1zu1Z69VdzszQae92utvZb7t7/Pw/f55ZJPv+eyLYZhGAIAADCRfv5uAAAAoLsRgAAAgOkQgAAAgOkQgAAAgOkQgAAAgOkQgAAAgOkQgAAAgOkE+ruBnqijo0NHjhzRwIEDZbFY/N0OAADoBMMwdOLECcXExKhfv3Ov8RCA3Dhy5IhiY2P93QYAAPDA4cOHddlll52zhgDkxsCBAyWdnsDQ0FA/dwMAADqjublZsbGxjt/j50IAcuPMaa/Q0FACEAAAvUxnLl/hImgAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6fg9A+fn5io+Pl9Vqlc1m044dO85ZX1ZWJpvNJqvVqoSEBBUUFLjUfP/995o1a5aio6NltVqVmJio4uJiXx0CAADoZfwagDZt2qS5c+cqOztb1dXVSktL04QJE1RXV+e2/sCBA5o4caLS0tJUXV2thQsXavbs2SoqKnLUtLa26rbbbtPBgwf1xz/+Ufv27dOaNWs0ZMiQ7josAADQw1kMwzD89cOvv/56jRkzRqtWrXKMJSYm6q677lJOTo5L/bx58/Tee++ptrbWMZaZmak9e/aooqJCklRQUKBXXnlFf/nLXxQUFNSpPlpaWtTS0uJ439zcrNjYWDU1NSk0NNTTwwMAAN2oublZYWFhnfr9HdhNPblobW1VVVWV5s+f7zSenp6u8vJyt9tUVFQoPT3daWz8+PEqLCzUqVOnFBQUpPfee08pKSmaNWuW3n33XV1yySW6//77NW/ePAUEBLjdb05OjhYvXuydAwMAwOSGzf/gvDUHl97eDZ2cnd9OgTU2Nqq9vV2RkZFO45GRkWpoaHC7TUNDg9v6trY2NTY2SpL279+vP/7xj2pvb1dxcbGefvpp/ed//qdefPHFs/ayYMECNTU1OV6HDx/u4tEBAICezG8rQGdYLBan94ZhuIydr/6n4x0dHbr00ku1evVqBQQEyGaz6ciRI3rllVf07LPPut1nSEiIQkJCunIYAACgF/FbAIqIiFBAQIDLas/Ro0ddVnnOiIqKclsfGBio8PBwSVJ0dLSCgoKcTnclJiaqoaFBra2tCg4O9vKRAACA3sZvp8CCg4Nls9lUWlrqNF5aWqrU1FS326SkpLjUl5SUKDk52XHB89ixY/W3v/1NHR0djpq//vWvio6OJvwAAABJfr4N3m63a+3atVq3bp1qa2uVlZWluro6ZWZmSjp9bc5DDz3kqM/MzNShQ4dkt9tVW1urdevWqbCwUE899ZSj5vHHH9exY8c0Z84c/fWvf9UHH3yg3/72t5o1a1a3Hx8AAOiZ/HoNUEZGho4dO6YlS5aovr5eSUlJKi4uVlxcnCSpvr7e6ZlA8fHxKi4uVlZWllauXKmYmBgtX75cU6ZMcdTExsaqpKREWVlZGjVqlIYMGaI5c+Zo3rx53X58AACgZ/Lrc4B6qgt5jgAAAHDmr9vgL+T3t9+/CgMAAKC7EYAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDp+D0A5efnKz4+XlarVTabTTt27DhnfVlZmWw2m6xWqxISElRQUOD0+RtvvCGLxeLy+vHHH315GAAAoBfxawDatGmT5s6dq+zsbFVXVystLU0TJkxQXV2d2/oDBw5o4sSJSktLU3V1tRYuXKjZs2erqKjIqS40NFT19fVOL6vV2h2HBAAAeoFAf/7w3NxcTZ8+XTNmzJAk5eXlacuWLVq1apVycnJc6gsKCjR06FDl5eVJkhITE1VZWally5ZpypQpjjqLxaKoqKhuOQYAAND7+G0FqLW1VVVVVUpPT3caT09PV3l5udttKioqXOrHjx+vyspKnTp1yjH2j3/8Q3Fxcbrssss0adIkVVdXn7OXlpYWNTc3O70AAEDf5bcA1NjYqPb2dkVGRjqNR0ZGqqGhwe02DQ0Nbuvb2trU2NgoSRo5cqTeeOMNvffee9qwYYOsVqvGjh2rr7/++qy95OTkKCwszPGKjY3t4tEBAICezO8XQVssFqf3hmG4jJ2v/qfjN9xwgx588EFdffXVSktL01tvvaXLL79cr7322ln3uWDBAjU1NTlehw8f9vRwAABAL+C3a4AiIiIUEBDgstpz9OhRl1WeM6KiotzWBwYGKjw83O02/fr107XXXnvOFaCQkBCFhIRc4BEAAIDeym8rQMHBwbLZbCotLXUaLy0tVWpqqtttUlJSXOpLSkqUnJysoKAgt9sYhqGamhpFR0d7p3EAANDr+fUUmN1u19q1a7Vu3TrV1tYqKytLdXV1yszMlHT61NRDDz3kqM/MzNShQ4dkt9tVW1urdevWqbCwUE899ZSjZvHixdqyZYv279+vmpoaTZ8+XTU1NY59AgAA+PU2+IyMDB07dkxLlixRfX29kpKSVFxcrLi4OElSfX290zOB4uPjVVxcrKysLK1cuVIxMTFavny50y3w33//vR577DE1NDQoLCxM11xzjbZv367rrruu248PAAD0TBbjzFXEcGhublZYWJiampoUGhrq73YAAOhVhs3/4Lw1B5fe7vWfeyG/v/1+FxgAAEB3IwABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADTIQABAADT8XsAys/PV3x8vKxWq2w2m3bs2HHO+rKyMtlsNlmtViUkJKigoOCstRs3bpTFYtFdd93l5a4BAEBv5tcAtGnTJs2dO1fZ2dmqrq5WWlqaJkyYoLq6Orf1Bw4c0MSJE5WWlqbq6motXLhQs2fPVlFRkUvtoUOH9NRTTyktLc3XhwEAAHoZvwag3NxcTZ8+XTNmzFBiYqLy8vIUGxurVatWua0vKCjQ0KFDlZeXp8TERM2YMUPTpk3TsmXLnOra29v1wAMPaPHixUpISOiOQwEAAL2I3wJQa2urqqqqlJ6e7jSenp6u8vJyt9tUVFS41I8fP16VlZU6deqUY2zJkiW65JJLNH369E710tLSoubmZqcXAADou/wWgBobG9Xe3q7IyEin8cjISDU0NLjdpqGhwW19W1ubGhsbJUm7du1SYWGh1qxZ0+lecnJyFBYW5njFxsZe4NEAAIDexO8XQVssFqf3hmG4jJ2v/sz4iRMn9OCDD2rNmjWKiIjodA8LFixQU1OT43X48OELOAIAANDbBPrrB0dERCggIMBltefo0aMuqzxnREVFua0PDAxUeHi4vvzySx08eFCTJ092fN7R0SFJCgwM1L59+zR8+HCX/YaEhCgkJKSrhwQAAHoJv60ABQcHy2azqbS01Gm8tLRUqampbrdJSUlxqS8pKVFycrKCgoI0cuRI7d27VzU1NY7XHXfcoZtvvlk1NTWc2gIAAJL8uAIkSXa7XVOnTlVycrJSUlK0evVq1dXVKTMzU9LpU1Pfffed1q9fL0nKzMzUihUrZLfb9eijj6qiokKFhYXasGGDJMlqtSopKcnpZwwaNEiSXMYBAIB5+TUAZWRk6NixY1qyZInq6+uVlJSk4uJixcXFSZLq6+udngkUHx+v4uJiZWVlaeXKlYqJidHy5cs1ZcoUfx0CAADohSzGmauI4dDc3KywsDA1NTUpNDTU3+0AANCrDJv/wXlrDi693es/90J+f/v9LjAAAIDu5tdTYACAnsdf//cOdCdWgAAAgOkQgAAAgOkQgAAAgOkQgAAAgOkQgAAAgOlwFxgA4IJxpxh6O49WgE6ePOntPgAAALqNRwEoMjJS06ZN086dO73dDwAAgM95FIA2bNigpqYm3XLLLbr88su1dOlSHTlyxNu9AQAA+IRHAWjy5MkqKirSkSNH9Pjjj2vDhg2Ki4vTpEmTtHnzZrW1tXm7TwAAAK/p0l1g4eHhysrK0p49e5Sbm6uPPvpI99xzj2JiYvTss8/qhx9+8FafAAAAXtOlu8AaGhq0fv16vf7666qrq9M999yj6dOn68iRI1q6dKk+/fRTlZSUeKtXAAAAr/AoAG3evFmvv/66tmzZoiuvvFKzZs3Sgw8+qEGDBjlqRo8erWuuucZbfQIAAHiNRwHokUce0a9//Wvt2rVL1157rduahIQEZWdnd6k5AAAAX/AoANXX1+tf/uVfzlnTv39/LVq0yKOmAAAAfMmji6AHDhyoo0ePuowfO3ZMAQEBXW4KAADAlzwKQIZhuB1vaWlRcHBwlxoCAADwtQs6BbZ8+XJJksVi0dq1azVgwADHZ+3t7dq+fbtGjhzp3Q4BAAC87IIC0Kuvvirp9ApQQUGB0+mu4OBgDRs2TAUFBd7tEAAAwMsuKAAdOHBAknTzzTdr8+bNGjx4sE+aAgAA8CWP7gLbunWrt/sA+rxh8z84b83Bpbd3QycAgE4HILvdrueff14XXXSR7Hb7OWtzc3O73BgAAICvdDoAVVdX69SpU44/n43FYul6VwAAAD7U6QD009NenAIDAAC9WZe+DR4AAKA36vQK0K9+9atO73Tz5s0eNQMAANAdOh2AwsLCfNkHAABAt+l0AHr99dd92QcAAEC34RogAABgOp1eARozZow+/vhjDR48WNdcc805b3ffvXu3V5oDAADwhU4HoDvvvFMhISGSpLvuustX/QAAfKgzTyQHzKDTAWjRokVu/wwAANDbePRdYGdUVlaqtrZWFotFiYmJstls3uoLAADAZzwKQN9++63uu+8+7dq1S4MGDZIkff/990pNTdWGDRsUGxvrzR4B0+ALUwGge3h0F9i0adN06tQp1dbW6vjx4zp+/Lhqa2tlGIamT5/u7R4BAAC8yqMVoB07dqi8vFxXXHGFY+yKK67Qa6+9prFjx3qtOQAAAF/waAVo6NChjm+G/6m2tjYNGTKky00BAAD4kkcB6OWXX9aTTz6pyspKGYYh6fQF0XPmzNGyZcu82iAAAIC3dfoU2ODBg50efnjy5Eldf/31Cgw8vYu2tjYFBgZq2rRpPCcIAAD0aJ0OQHl5eT5sAwAAoPt0OgA9/PDDvuwDAACg23TpQYiS9H//938uF0SHhoZ2dbcAAAA+49FF0CdPntQTTzyhSy+9VAMGDNDgwYOdXgAAAD2ZRwHoP/7jP/TJJ58oPz9fISEhWrt2rRYvXqyYmBitX7/e2z0CAAB4lUenwN5//32tX79eN910k6ZNm6a0tDSNGDFCcXFxevPNN/XAAw94u08AAACv8WgF6Pjx44qPj5d0+nqf48ePS5JuvPFGbd++/YL2lZ+fr/j4eFmtVtlsNu3YseOc9WVlZbLZbLJarUpISFBBQYHT55s3b1ZycrIGDRqkiy66SKNHj9bvf//7C+oJAAD0bR4FoISEBB08eFCSdOWVV+qtt96SdHpl6MyXo3bGpk2bNHfuXGVnZ6u6ulppaWmaMGGC6urq3NYfOHBAEydOVFpamqqrq7Vw4ULNnj1bRUVFjpqLL75Y2dnZqqio0BdffKFHHnlEjzzyiLZs2eLJoQIAgD7IowD0yCOPaM+ePZKkBQsWOK4FysrK0r//+793ej+5ubmaPn26ZsyYocTEROXl5Sk2NlarVq1yW19QUKChQ4cqLy9PiYmJmjFjhqZNm+b09OmbbrpJd999txITEzV8+HDNmTNHo0aN0s6dOz05VAAA0Ad5dA1QVlaW488333yzamtrVVVVpeHDh+vqq6/u1D5aW1tVVVWl+fPnO42np6ervLzc7TYVFRVKT093Ghs/frwKCwt16tQpBQUFOX1mGIY++eQT7du3Ty+99NJZe2lpaVFLS4vjfXNzc6eOAQAA9E5dfg6QJMXFxSkuLu6CtmlsbFR7e7siIyOdxiMjI9XQ0OB2m4aGBrf1bW1tamxsVHR0tCSpqalJQ4YMUUtLiwICApSfn6/bbrvtrL3k5ORo8eLFF9Q/AADovTw6BSZJH3/8sSZNmqThw4drxIgRmjRpkj766KML3s9Pv19MOr1q889j56v/5/GBAweqpqZGn3/+uV588UXZ7XZt27btrPtcsGCBmpqaHK/Dhw9f8HEAAIDew6MVoBUrVigrK0v33HOP5syZI0n69NNPNXHiROXm5uqJJ5447z4iIiIUEBDgstpz9OhRl1WeM6KiotzWBwYGKjw83DHWr18/jRgxQpI0evRo1dbWKicnRzfddJPb/YaEhCgkJOS8PQOAvwyb/8F5aw4uvb0bOgH6Bo8CUE5Ojl599VWnoDN79myNHTtWL774YqcCUHBwsGw2m0pLS3X33Xc7xktLS3XnnXe63SYlJUXvv/++01hJSYmSk5Ndrv/5KcMwnK7xAQAA5ubRKbDm5mb98pe/dBlPT0+/oAuI7Xa71q5dq3Xr1qm2tlZZWVmqq6tTZmampNOnph566CFHfWZmpg4dOiS73a7a2lqtW7dOhYWFeuqppxw1OTk5Ki0t1f79+/WXv/xFubm5Wr9+vR588EFPDhUAAPRBHq0A3XHHHXr77bddbnl/9913NXny5E7vJyMjQ8eOHdOSJUtUX1+vpKQkFRcXOy6orq+vd3omUHx8vIqLi5WVlaWVK1cqJiZGy5cv15QpUxw1J0+e1MyZM/Xtt9+qf//+GjlypP7whz8oIyPDk0MFAAB9UKcD0PLlyx1/TkxM1Isvvqht27YpJSVF0ulrgHbt2qXf/OY3F9TAzJkzNXPmTLefvfHGGy5j48aN0+7du8+6vxdeeEEvvPDCBfUAAADMpdMB6NVXX3V6P3jwYH311Vf66quvHGODBg3SunXr9PTTT3uvQwC4AFws3HPw3wI9WacD0IEDB3zZBwAAQLfx+DlAZxiG4XgWDwAAQG/gcQBav369rrrqKvXv31/9+/fXqFGj+NZ1AADQK3h0F1hubq6eeeYZPfHEExo7dqwMw9CuXbuUmZmpxsZGp+8KAwBv6cw1JQDQGR4FoNdee02rVq1yekbPnXfeqZ///Od67rnnCEAAAKBH8ygA1dfXKzU11WU8NTVV9fX1XW4KAHDhWCEDOs+ja4BGjBiht956y2V806ZN+tnPftblpgAAAHzJoxWgxYsXKyMjQ9u3b9fYsWNlsVi0c+dOffzxx26DEQAAQE/i0QrQlClT9NlnnykiIkLvvPOONm/erIiICH322WdOX2wKAADQE13wCtCpU6f02GOP6ZlnntEf/vAHX/QEAADgUxe8AhQUFKS3337bF70AAAB0C49Ogd1999165513vNwKAABA9/DoIugRI0bo+eefV3l5uWw2my666CKnz2fPnu2V5gAAAHzBowC0du1aDRo0SFVVVaqqqnL6zGKxEIAAAECP5lEA+uk3w5/5IlSLxeKdjgAAAHzM4y9DLSwsVFJSkqxWq6xWq5KSkrR27Vpv9gYAAOATHq0APfPMM3r11Vf15JNPKiUlRZJUUVGhrKwsHTx4UC+88IJXmwQAAPAmjwLQqlWrtGbNGt13332OsTvuuEOjRo3Sk08+SQACAAA9mkenwNrb25WcnOwybrPZ1NbW1uWmAAAAfMmjAPTggw9q1apVLuOrV6/WAw880OWmAAAAfMmjU2DS6YugS0pKdMMNN0iSPv30Ux0+fFgPPfSQ7Ha7oy43N7frXQI93LD5H/i7BQDABfAoAP35z3/WmDFjJEnffPONJOmSSy7RJZdcoj//+c+OOm6NBwAAPZFHAWjr1q3e7gNAJ3lrteng0tu9sh+cW2f+e/HfAuh+Hj8HCAAAoLfy+BogAPCm3ngdVW/sGcBprAABAADTYQUIgOlwXQ4AVoAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpEIAAAIDpBPq7AQCAeQ2b/8F5aw4uvb0bOoHZsAIEAABMx+8BKD8/X/Hx8bJarbLZbNqxY8c568vKymSz2WS1WpWQkKCCggKnz9esWaO0tDQNHjxYgwcP1q233qrPPvvMl4cA9ErD5n9w3hcA9FV+DUCbNm3S3LlzlZ2drerqaqWlpWnChAmqq6tzW3/gwAFNnDhRaWlpqq6u1sKFCzV79mwVFRU5arZt26b77rtPW7duVUVFhYYOHar09HR999133XVYAACgh/NrAMrNzdX06dM1Y8YMJSYmKi8vT7GxsVq1apXb+oKCAg0dOlR5eXlKTEzUjBkzNG3aNC1btsxR8+abb2rmzJkaPXq0Ro4cqTVr1qijo0Mff/xxdx0WAADo4fwWgFpbW1VVVaX09HSn8fT0dJWXl7vdpqKiwqV+/Pjxqqys1KlTp9xu88MPP+jUqVO6+OKLz9pLS0uLmpubnV4AAKDv8ttdYI2NjWpvb1dkZKTTeGRkpBoaGtxu09DQ4La+ra1NjY2Nio6Odtlm/vz5GjJkiG699daz9pKTk6PFixd7cBRA38YdOgD6Kr9fBG2xWJzeG4bhMna+enfjkvTyyy9rw4YN2rx5s6xW61n3uWDBAjU1NTlehw8fvpBDAAAAvYzfVoAiIiIUEBDgstpz9OhRl1WeM6KiotzWBwYGKjw83Gl82bJl+u1vf6uPPvpIo0aNOmcvISEhCgkJ8eAoALBKBKA38tsKUHBwsGw2m0pLS53GS0tLlZqa6nablJQUl/qSkhIlJycrKCjIMfbKK6/o+eef14cffqjk5GTvNw8AAHo1v54Cs9vtWrt2rdatW6fa2lplZWWprq5OmZmZkk6fmnrooYcc9ZmZmTp06JDsdrtqa2u1bt06FRYW6qmnnnLUvPzyy3r66ae1bt06DRs2TA0NDWpoaNA//vGPbj8+AADQM/n1qzAyMjJ07NgxLVmyRPX19UpKSlJxcbHi4uIkSfX19U7PBIqPj1dxcbGysrK0cuVKxcTEaPny5ZoyZYqjJj8/X62trbrnnnucftaiRYv03HPPdctxAQCAns3v3wU2c+ZMzZw50+1nb7zxhsvYuHHjtHv37rPu7+DBg17qDICZ8SRsoG/z+11gAAAA3c3vK0BAT8dKAHyNv2NA92MFCAAAmA4rQAAA0/DWc6t4/lXvxwoQAAAwHVaAAAA9WmevkWLFBReCFSAAAGA6BCAAAGA6nAID4HPc5g2gp2EFCAAAmA4BCAAAmA4BCAAAmA7XAAEA8BPeumbNW/vh9n7fYAUIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYTqC/GwAAwBuGzf/A3y2gF2EFCAAAmA4BCAAAmA4BCAAAmA7XAKFX4lw/ALPozL93B5fe3g2d9C2sAAEAANMhAAEAANMhAAEAANPhGqA+jnPHAAC4YgUIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDgEIAACYDk+CBgDABPhmAGesAAEAANMhAAEAANPxewDKz89XfHy8rFarbDabduzYcc76srIy2Ww2Wa1WJSQkqKCgwOnzL7/8UlOmTNGwYcNksViUl5fnw+4BAEBv5NcAtGnTJs2dO1fZ2dmqrq5WWlqaJkyYoLq6Orf1Bw4c0MSJE5WWlqbq6motXLhQs2fPVlFRkaPmhx9+UEJCgpYuXaqoqKjuOhQAANCL+PUi6NzcXE2fPl0zZsyQJOXl5WnLli1atWqVcnJyXOoLCgo0dOhQx6pOYmKiKisrtWzZMk2ZMkWSdO211+raa6+VJM2fP79TfbS0tKilpcXxvrm5uSuHBQAAeji/rQC1traqqqpK6enpTuPp6ekqLy93u01FRYVL/fjx41VZWalTp0553EtOTo7CwsIcr9jYWI/3BQAAej6/rQA1Njaqvb1dkZGRTuORkZFqaGhwu01DQ4Pb+ra2NjU2Nio6OtqjXhYsWCC73e5439zcTAjyo87cqgkAQFf4/TlAFovF6b1hGC5j56t3N34hQkJCFBIS4vH2AACgd/HbKbCIiAgFBAS4rPYcPXrUZZXnjKioKLf1gYGBCg8P91mvAACgb/FbAAoODpbNZlNpaanTeGlpqVJTU91uk5KS4lJfUlKi5ORkBQUF+axXAADQt/j1FJjdbtfUqVOVnJyslJQUrV69WnV1dcrMzJR0+tqc7777TuvXr5ckZWZmasWKFbLb7Xr00UdVUVGhwsJCbdiwwbHP1tZWffXVV44/f/fdd6qpqdGAAQM0YsSI7j9IH+JaGQAAPOPXAJSRkaFjx45pyZIlqq+vV1JSkoqLixUXFydJqq+vd3omUHx8vIqLi5WVlaWVK1cqJiZGy5cvd9wCL0lHjhzRNddc43i/bNkyLVu2TOPGjdO2bdu67dgAAEDP5feLoGfOnKmZM2e6/eyNN95wGRs3bpx279591v0NGzbMcWE0eh5WrQAAPYHfvwoDAACguxGAAACA6RCAAACA6fj9GiD4X2euyzm49Hav7AcAgJ6AFSAAAGA6BCAAAGA6BCAAAGA6BCAAAGA6BCAAAGA63AUGAEAvx124F44VIAAAYDqsAAEAAEneey5cb0AAQqewvAoA6Es4BQYAAEyHAAQAAEyHAAQAAEyHAAQAAEyHAAQAAEyHAAQAAEyHAAQAAEyH5wABAIBO6yvPhWMFCAAAmA4BCAAAmA4BCAAAmA4BCAAAmA4BCAAAmA4BCAAAmA4BCAAAmA7PAeqh+spzFgAA6IlYAQIAAKZDAAIAAKZDAAIAAKZDAAIAAKZDAAIAAKZDAAIAAKbDbfB+wC3uAAD4FytAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdPwegPLz8xUfHy+r1SqbzaYdO3acs76srEw2m01Wq1UJCQkqKChwqSkqKtKVV16pkJAQXXnllXr77bd91T4AAOiF/BqANm3apLlz5yo7O1vV1dVKS0vThAkTVFdX57b+wIEDmjhxotLS0lRdXa2FCxdq9uzZKioqctRUVFQoIyNDU6dO1Z49ezR16lTde++9+tOf/tRdhwUAAHo4i2EYhr9++PXXX68xY8Zo1apVjrHExETdddddysnJcamfN2+e3nvvPdXW1jrGMjMztWfPHlVUVEiSMjIy1NzcrP/5n/9x1Pzyl7/U4MGDtWHDhk711dzcrLCwMDU1NSk0NNTTwzsrvg0eAGB2B5fe7vV9Xsjv70Cv//ROam1tVVVVlebPn+80np6ervLycrfbVFRUKD093Wls/PjxKiws1KlTpxQUFKSKigplZWW51OTl5Z21l5aWFrW0tDjeNzU1STo9kb7Q0fKDT/YLAEBv4YvfsWf22Zm1Hb8FoMbGRrW3tysyMtJpPDIyUg0NDW63aWhocFvf1tamxsZGRUdHn7XmbPuUpJycHC1evNhlPDY2trOHAwAALkBYnu/2feLECYWFhZ2zxm8B6AyLxeL03jAMl7Hz1f/z+IXuc8GCBbLb7Y73HR0dOn78uMLDw8+53T9rbm5WbGysDh8+7JNTZ2bH/Poec+xbzK/vMce+1dPn1zAMnThxQjExMeet9VsAioiIUEBAgMvKzNGjR11WcM6IiopyWx8YGKjw8PBz1pxtn5IUEhKikJAQp7FBgwZ19lBchIaG9si/GH0F8+t7zLFvMb++xxz7Vk+e3/Ot/Jzht7vAgoODZbPZVFpa6jReWlqq1NRUt9ukpKS41JeUlCg5OVlBQUHnrDnbPgEAgPn49RSY3W7X1KlTlZycrJSUFK1evVp1dXXKzMyUdPrU1Hfffaf169dLOn3H14oVK2S32/Xoo4+qoqJChYWFTnd3zZkzR7/4xS/00ksv6c4779S7776rjz76SDt37vTLMQIAgJ7HrwEoIyNDx44d05IlS1RfX6+kpCQVFxcrLi5OklRfX+/0TKD4+HgVFxcrKytLK1euVExMjJYvX64pU6Y4alJTU7Vx40Y9/fTTeuaZZzR8+HBt2rRJ119/vc+PJyQkRIsWLXI5nQbvYH59jzn2LebX95hj3+pL8+vX5wABAAD4g9+/CgMAAKC7EYAAAIDpEIAAAIDpEIAAAIDpEIC8JD8/X/Hx8bJarbLZbNqxY4e/W+q1tm/frsmTJysmJkYWi0XvvPOO0+eGYei5555TTEyM+vfvr5tuuklffvmlf5rthXJycnTttddq4MCBuvTSS3XXXXdp3759TjXMsedWrVqlUaNGOR4Ul5KS4vTlzMytd+Xk5MhisWju3LmOMea4a5577jlZLBanV1RUlOPzvjK/BCAv2LRpk+bOnavs7GxVV1crLS1NEyZMcLqFH5138uRJXX311VqxYoXbz19++WXl5uZqxYoV+vzzzxUVFaXbbrtNJ06c6OZOe6eysjLNmjVLn376qUpLS9XW1qb09HSdPHnSUcMce+6yyy7T0qVLVVlZqcrKSv3rv/6r7rzzTscvCObWez7//HOtXr1ao0aNchpnjrvu5z//uerr6x2vvXv3Oj7rM/NroMuuu+46IzMz02ls5MiRxvz58/3UUd8hyXj77bcd7zs6OoyoqChj6dKljrEff/zRCAsLMwoKCvzQYe939OhRQ5JRVlZmGAZz7AuDBw821q5dy9x60YkTJ4yf/exnRmlpqTFu3Dhjzpw5hmHw99cbFi1aZFx99dVuP+tL88sKUBe1traqqqpK6enpTuPp6ekqLy/3U1d914EDB9TQ0OA03yEhIRo3bhzz7aGmpiZJ0sUXXyyJOfam9vZ2bdy4USdPnlRKSgpz60WzZs3S7bffrltvvdVpnDn2jq+//loxMTGKj4/Xr3/9a+3fv19S35pfv38bfG/X2Nio9vZ2ly9bjYyMdPlSVnTdmTl1N9+HDh3yR0u9mmEYstvtuvHGG5WUlCSJOfaGvXv3KiUlRT/++KMGDBigt99+W1deeaXjFwRz2zUbN27U7t279fnnn7t8xt/frrv++uu1fv16XX755fr73/+uF154Qampqfryyy/71PwSgLzEYrE4vTcMw2UM3sN8e8cTTzyhL774wu135THHnrviiitUU1Oj77//XkVFRXr44YdVVlbm+Jy59dzhw4c1Z84clZSUyGq1nrWOOfbchAkTHH++6qqrlJKSouHDh+t3v/udbrjhBkl9Y345BdZFERERCggIcFntOXr0qEtCRteduROB+e66J598Uu+99562bt2qyy67zDHOHHddcHCwRowYoeTkZOXk5Ojqq6/Wf/3XfzG3XlBVVaWjR4/KZrMpMDBQgYGBKisr0/LlyxUYGOiYR+bYey666CJdddVV+vrrr/vU32ECUBcFBwfLZrOptLTUaby0tFSpqal+6qrvio+PV1RUlNN8t7a2qqysjPnuJMMw9MQTT2jz5s365JNPFB8f7/Q5c+x9hmGopaWFufWCW265RXv37lVNTY3jlZycrAceeEA1NTVKSEhgjr2spaVFtbW1io6O7lt/h/12+XUfsnHjRiMoKMgoLCw0vvrqK2Pu3LnGRRddZBw8eNDfrfVKJ06cMKqrq43q6mpDkpGbm2tUV1cbhw4dMgzDMJYuXWqEhYUZmzdvNvbu3Wvcd999RnR0tNHc3OznznuHxx9/3AgLCzO2bdtm1NfXO14//PCDo4Y59tyCBQuM7du3GwcOHDC++OILY+HChUa/fv2MkpISwzCYW1/46V1ghsEcd9VvfvMbY9u2bcb+/fuNTz/91Jg0aZIxcOBAx++0vjK/BCAvWblypREXF2cEBwcbY8aMcdxSjAu3detWQ5LL6+GHHzYM4/RtmIsWLTKioqKMkJAQ4xe/+IWxd+9e/zbdi7ibW0nG66+/7qhhjj03bdo0x78Fl1xyiXHLLbc4wo9hMLe+8M8BiDnumoyMDCM6OtoICgoyYmJijF/96lfGl19+6fi8r8yvxTAMwz9rTwAAAP7BNUAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAAAMB0CEAA+oQPP/xQN954owYNGqTw8HBNmjRJ33zzjePz8vJyjR49WlarVcnJyXrnnXdksVhUU1PjqPnqq680ceJEDRgwQJGRkZo6daoaGxv9cDQAfI0ABKBPOHnypOx2uz7//HN9/PHH6tevn+6++251dHToxIkTmjx5sq666irt3r1bzz//vObNm+e0fX19vcaNG6fRo0ersrJSH374of7+97/r3nvv9dMRAfAlvg0eQJ/0v//7v7r00ku1d+9e7dy5U08//bS+/fZbWa1WSdLatWv16KOPqrq6WqNHj9azzz6rP/3pT9qyZYtjH99++61iY2O1b98+XX755f46FAA+wAoQgD7hm2++0f3336+EhASFhoYqPj5eklRXV6d9+/Zp1KhRjvAjSdddd53T9lVVVdq6dasGDBjgeI0cOdKxbwB9S6C/GwAAb5g8ebJiY2O1Zs0axcTEqKOjQ0lJSWptbZVhGLJYLE71/7z43dHRocmTJ+ull15y2Xd0dLRPewfQ/QhAAHq9Y8eOqba2Vv/93/+ttLQ0SdLOnTsdn48cOVJvvvmmWlpaFBISIkmqrKx02seYMWNUVFSkYcOGKTCQfxqBvo5TYAB6vcGDBys8PFyrV6/W3/72N33yySey2+2Oz++//351dHToscceU21trbZs2aJly5ZJkmNlaNasWTp+/Ljuu+8+ffbZZ9q/f79KSko0bdo0tbe3++W4APgOAQhAr9evXz9t3LhRVVVVSkpKUlZWll555RXH56GhoXr//fdVU1Oj0aNHKzs7W88++6wkOa4LiomJ0a5du9Te3q7x48crKSlJc+bMUVhYmPr1459KoK/hLjAApvTmm2/qkUceUVNTk/r37+/vdgB0M050AzCF9evXKyEhQUOGDNGePXs0b9483XvvvYQfwKQIQABMoaGhQc8++6waGhoUHR2tf/u3f9OLL77o77YA+AmnwAAAgOlwZR8AADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADAdAhAAADCd/wfi53kXi9WxzwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Lets plot some of that data with a module called matplotlib. We will learn more about this module\n",
    "# later in the course.  Here we are just using it to plot a histogram of the housing median age.\n",
    "\n",
    "# First, we need to import the module, and here we are using the alias \"plt\" for the function \"pyplot\"\n",
    "# from the module \"matplotlib\". \n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#pull the values out of the pandas array (stored as a numpy array)\n",
    "housing_data_array = housing_dataframe['housing_median_age'].values\n",
    "\n",
    "# Now to make the histogram and add labels.\n",
    "plt.hist(housing_data_array,50,density=True)\n",
    "plt.xlabel('age')\n",
    "plt.ylabel('probability')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The attributes of the pandas dataframe are:\n",
      "['T', '_AXIS_LEN', '_AXIS_ORDERS', '_AXIS_TO_AXIS_NUMBER', '_HANDLED_TYPES', '__abs__', '__add__', '__and__', '__annotations__', '__array__', '__array_priority__', '__array_ufunc__', '__arrow_c_stream__', '__bool__', '__class__', '__contains__', '__copy__', '__dataframe__', '__dataframe_consortium_standard__', '__deepcopy__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__divmod__', '__doc__', '__eq__', '__finalize__', '__floordiv__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__iadd__', '__iand__', '__ifloordiv__', '__imod__', '__imul__', '__init__', '__init_subclass__', '__invert__', '__ior__', '__ipow__', '__isub__', '__iter__', '__itruediv__', '__ixor__', '__le__', '__len__', '__lt__', '__matmul__', '__mod__', '__module__', '__mul__', '__ne__', '__neg__', '__new__', '__nonzero__', '__or__', '__pandas_priority__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rmatmul__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__setitem__', '__setstate__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__weakref__', '__xor__', '_accessors', '_accum_func', '_agg_examples_doc', '_agg_see_also_doc', '_align_for_op', '_align_frame', '_align_series', '_append', '_arith_method', '_arith_method_with_reindex', '_as_manager', '_attrs', '_box_col_values', '_can_fast_transpose', '_check_inplace_and_allows_duplicate_labels', '_check_is_chained_assignment_possible', '_check_label_or_level_ambiguity', '_check_setitem_copy', '_clear_item_cache', '_clip_with_one_bound', '_clip_with_scalar', '_cmp_method', '_combine_frame', '_consolidate', '_consolidate_inplace', '_construct_axes_dict', '_construct_result', '_constructor', '_constructor_from_mgr', '_constructor_sliced', '_constructor_sliced_from_mgr', '_create_data_for_split_and_tight_to_dict', '_data', '_deprecate_downcast', '_dir_additions', '_dir_deletions', '_dispatch_frame_op', '_drop_axis', '_drop_labels_or_levels', '_ensure_valid_index', '_find_valid_index', '_flags', '_flex_arith_method', '_flex_cmp_method', '_from_arrays', '_from_mgr', '_get_agg_axis', '_get_axis', '_get_axis_name', '_get_axis_number', '_get_axis_resolvers', '_get_block_manager_axis', '_get_bool_data', '_get_cleaned_column_resolvers', '_get_column_array', '_get_index_resolvers', '_get_item_cache', '_get_label_or_level_values', '_get_numeric_data', '_get_value', '_get_values_for_csv', '_getitem_bool_array', '_getitem_multilevel', '_getitem_nocopy', '_getitem_slice', '_gotitem', '_hidden_attrs', '_indexed_same', '_info_axis', '_info_axis_name', '_info_axis_number', '_info_repr', '_init_mgr', '_inplace_method', '_internal_names', '_internal_names_set', '_is_copy', '_is_homogeneous_type', '_is_label_or_level_reference', '_is_label_reference', '_is_level_reference', '_is_mixed_type', '_is_view', '_is_view_after_cow_rules', '_iset_item', '_iset_item_mgr', '_iset_not_inplace', '_item_cache', '_iter_column_arrays', '_ixs', '_logical_func', '_logical_method', '_maybe_align_series_as_frame', '_maybe_cache_changed', '_maybe_update_cacher', '_metadata', '_mgr', '_min_count_stat_function', '_needs_reindex_multi', '_pad_or_backfill', '_protect_consolidate', '_reduce', '_reduce_axis1', '_reindex_axes', '_reindex_multi', '_reindex_with_indexers', '_rename', '_replace_columnwise', '_repr_data_resource_', '_repr_fits_horizontal_', '_repr_fits_vertical_', '_repr_html_', '_repr_latex_', '_reset_cache', '_reset_cacher', '_sanitize_column', '_series', '_set_axis', '_set_axis_name', '_set_axis_nocheck', '_set_is_copy', '_set_item', '_set_item_frame_value', '_set_item_mgr', '_set_value', '_setitem_array', '_setitem_frame', '_setitem_slice', '_shift_with_freq', '_should_reindex_frame_op', '_slice', '_stat_function', '_stat_function_ddof', '_take_with_is_copy', '_to_dict_of_blocks', '_to_latex_via_styler', '_typ', '_update_inplace', '_validate_dtype', '_values', '_where', 'abs', 'add', 'add_prefix', 'add_suffix', 'agg', 'aggregate', 'align', 'all', 'any', 'apply', 'applymap', 'asfreq', 'asof', 'assign', 'astype', 'at', 'at_time', 'attrs', 'axes', 'backfill', 'between_time', 'bfill', 'bool', 'boxplot', 'clip', 'columns', 'combine', 'combine_first', 'compare', 'convert_dtypes', 'copy', 'corr', 'corrwith', 'count', 'cov', 'cummax', 'cummin', 'cumprod', 'cumsum', 'describe', 'diff', 'div', 'divide', 'dot', 'drop', 'drop_duplicates', 'droplevel', 'dropna', 'dtypes', 'duplicated', 'empty', 'eq', 'equals', 'eval', 'ewm', 'expanding', 'explode', 'ffill', 'fillna', 'filter', 'first', 'first_valid_index', 'flags', 'floordiv', 'from_dict', 'from_records', 'ge', 'get', 'groupby', 'gt', 'head', 'hist', 'households', 'housing_median_age', 'iat', 'idxmax', 'idxmin', 'iloc', 'index', 'infer_objects', 'info', 'insert', 'interpolate', 'isetitem', 'isin', 'isna', 'isnull', 'items', 'iterrows', 'itertuples', 'join', 'keys', 'kurt', 'kurtosis', 'last', 'last_valid_index', 'latitude', 'le', 'loc', 'longitude', 'lt', 'map', 'mask', 'max', 'mean', 'median', 'median_house_value', 'median_income', 'melt', 'memory_usage', 'merge', 'min', 'mod', 'mode', 'mul', 'multiply', 'ndim', 'ne', 'nlargest', 'notna', 'notnull', 'nsmallest', 'nunique', 'pad', 'pct_change', 'pipe', 'pivot', 'pivot_table', 'plot', 'pop', 'population', 'pow', 'prod', 'product', 'quantile', 'query', 'radd', 'rank', 'rdiv', 'reindex', 'reindex_like', 'rename', 'rename_axis', 'reorder_levels', 'replace', 'resample', 'reset_index', 'rfloordiv', 'rmod', 'rmul', 'rolling', 'round', 'rpow', 'rsub', 'rtruediv', 'sample', 'select_dtypes', 'sem', 'set_axis', 'set_flags', 'set_index', 'shape', 'shift', 'size', 'skew', 'sort_index', 'sort_values', 'squeeze', 'stack', 'std', 'style', 'sub', 'subtract', 'sum', 'swapaxes', 'swaplevel', 'tail', 'take', 'to_clipboard', 'to_csv', 'to_dict', 'to_excel', 'to_feather', 'to_gbq', 'to_hdf', 'to_html', 'to_json', 'to_latex', 'to_markdown', 'to_numpy', 'to_orc', 'to_parquet', 'to_period', 'to_pickle', 'to_records', 'to_sql', 'to_stata', 'to_string', 'to_timestamp', 'to_xarray', 'to_xml', 'total_bedrooms', 'total_rooms', 'transform', 'transpose', 'truediv', 'truncate', 'tz_convert', 'tz_localize', 'unstack', 'update', 'value_counts', 'values', 'var', 'where', 'xs']\n"
     ]
    }
   ],
   "source": [
    "# What are the atributes of the a pandas dataframe?\n",
    "# There are a lot!  We can see them by using the dir command.\n",
    "print('The attributes of the pandas dataframe are:')\n",
    "print(dir(housing_dataframe))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The callable functions of the pandas dataframe are:\n",
      "['abs', 'add', 'add_prefix', 'add_suffix', 'agg', 'aggregate', 'align', 'all', 'any', 'apply', 'applymap', 'asfreq', 'asof', 'assign', 'astype', 'at_time', 'backfill', 'between_time', 'bfill', 'bool', 'boxplot', 'clip', 'combine', 'combine_first', 'compare', 'convert_dtypes', 'copy', 'corr', 'corrwith', 'count', 'cov', 'cummax', 'cummin', 'cumprod', 'cumsum', 'describe', 'diff', 'div', 'divide', 'dot', 'drop', 'drop_duplicates', 'droplevel', 'dropna', 'duplicated', 'eq', 'equals', 'eval', 'ewm', 'expanding', 'explode', 'ffill', 'fillna', 'filter', 'first', 'first_valid_index', 'floordiv', 'from_dict', 'from_records', 'ge', 'get', 'groupby', 'gt', 'head', 'hist', 'idxmax', 'idxmin', 'iloc', 'infer_objects', 'info', 'insert', 'interpolate', 'isetitem', 'isin', 'isna', 'isnull', 'items', 'iterrows', 'itertuples', 'join', 'keys', 'kurt', 'kurtosis', 'last', 'last_valid_index', 'le', 'loc', 'lt', 'map', 'mask', 'max', 'mean', 'median', 'melt', 'memory_usage', 'merge', 'min', 'mod', 'mode', 'mul', 'multiply', 'ne', 'nlargest', 'notna', 'notnull', 'nsmallest', 'nunique', 'pad', 'pct_change', 'pipe', 'pivot', 'pivot_table', 'plot', 'pop', 'pow', 'prod', 'product', 'quantile', 'query', 'radd', 'rank', 'rdiv', 'reindex', 'reindex_like', 'rename', 'rename_axis', 'reorder_levels', 'replace', 'resample', 'reset_index', 'rfloordiv', 'rmod', 'rmul', 'rolling', 'round', 'rpow', 'rsub', 'rtruediv', 'sample', 'select_dtypes', 'sem', 'set_axis', 'set_flags', 'set_index', 'shift', 'skew', 'sort_index', 'sort_values', 'squeeze', 'stack', 'std', 'sub', 'subtract', 'sum', 'swapaxes', 'swaplevel', 'tail', 'take', 'to_clipboard', 'to_csv', 'to_dict', 'to_excel', 'to_feather', 'to_gbq', 'to_hdf', 'to_html', 'to_json', 'to_latex', 'to_markdown', 'to_numpy', 'to_orc', 'to_parquet', 'to_period', 'to_pickle', 'to_records', 'to_sql', 'to_stata', 'to_string', 'to_timestamp', 'to_xarray', 'to_xml', 'transform', 'transpose', 'truediv', 'truncate', 'tz_convert', 'tz_localize', 'unstack', 'update', 'value_counts', 'var', 'where', 'xs']\n"
     ]
    }
   ],
   "source": [
    "# Let's just look at the functions we can use on the dataframe.\n",
    "print('')\n",
    "print('The callable functions of the pandas dataframe are:')\n",
    "print([x for x in dir(housing_dataframe) if '__' not in x and x[0] != '_' and callable(getattr(housing_dataframe, x))])\n",
    "# What was that list comprehension doing?  It was looking at all of the attributes of the dataframe\n",
    "# and only selecting the ones that are callable (i.e., functions) and that do not start with an\n",
    "# underscore.  The ones that start with an underscore are private functions that are not meant to be\n",
    "# used by the user."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "8rvGU7XjOKlE"
   },
   "source": [
    "Here are a few common Pandas functions and what they do.\n",
    "| Attribute    | Description |\n",
    "| ----------- | ------------ |\n",
    "| .values()|  get a numpy array of all values within the dataframe    |  \n",
    "| .head()|  print out the top 5 rows    |  \n",
    "| .tail()|  print out the bottom 5 rows    |  \n",
    "| .iloc()|  lets you use indexes like the dataframe is numpy    |  \n",
    "| .columns()|  return out all the headers of the columns in the dataframe |  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "5ZFozGpmOrAp",
    "outputId": "cd5784e0-7714-49bb-c4d4-60eef6771ed8",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Try out some of these on the dataframe\n",
    "housing_dataframe.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if you want to know what another function does, you can use the help function\n",
    "help(housing_dataframe.aggregate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IpMOTILBMzU7"
   },
   "source": [
    "### **3.B) Writing a csv file with Pandas**\n",
    "\n",
    "Lets write some simple CSVs using pandas. Pandas accepts multiple formats as long as indexes are provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "C6WqFeNUMm0I",
    "outputId": "850fe68a-da77-4540-f1e4-83e3510a2e75",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in memory:\n",
      "       column 1 column 2 last col\n",
      "row 1         1        1        a\n",
      "row 2         2        2        b\n",
      "row 3         3        3        c\n",
      "row 4         4        4        d\n",
      "\n",
      "on disk:\n",
      "blank,column 1,column 2,last col\n",
      "row 1,1,1,a\n",
      "row 2,2,2,b\n",
      "row 3,3,3,c\n",
      "row 4,4,4,d\n"
     ]
    }
   ],
   "source": [
    "# The easiest way to write a new dataframe is to generate a dictionary of your data\n",
    "# data keyed by its column header\n",
    "new_dataframe_dict = {'column 1': [1,2,3,4],\n",
    "                      'column 2': ['1','2','3','4'],\n",
    "                      'last col': ['a','b','c','d']}\n",
    "\n",
    "# optional row labels, we have 4 entries for each column, so we need 4 labels\n",
    "row_labels = ['row 1', 'row 2', 'row 3', 'row 4']\n",
    "\n",
    "# Generate our new pandas dataframe\n",
    "new_df = pd.DataFrame(new_dataframe_dict, index=row_labels)\n",
    "print('in memory:')\n",
    "print(new_df.head())\n",
    "\n",
    "# We can write this data frame to a .csv for easy of access in multiple softwares\n",
    "print('')\n",
    "new_df.to_csv('test_csv.csv',  index_label='blank')\n",
    "print('on disk:')\n",
    "!head test_csv.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "o2gOEScwRMdi"
   },
   "source": [
    "### **3.C) Creating a dataframe from a numpy array**\n",
    "\n",
    "This is as simple as passing the array and the correct number of column headers into DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "id": "isUXp7hjPm-L",
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "random_array = np.random.randint(0,1000,size=(5,100)).T\n",
    "headers = ['col'+str(x) for x in range(0,5)]\n",
    "\n",
    "new_df = pd.DataFrame(random_array,columns=headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "Tvdr21iVQ1bw",
    "outputId": "27ae1765-96fc-4347-9dce-2206c2d72495",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col0</th>\n",
       "      <th>col1</th>\n",
       "      <th>col2</th>\n",
       "      <th>col3</th>\n",
       "      <th>col4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>938</td>\n",
       "      <td>668</td>\n",
       "      <td>550</td>\n",
       "      <td>140</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>932</td>\n",
       "      <td>978</td>\n",
       "      <td>730</td>\n",
       "      <td>919</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>76</td>\n",
       "      <td>839</td>\n",
       "      <td>747</td>\n",
       "      <td>169</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>624</td>\n",
       "      <td>579</td>\n",
       "      <td>770</td>\n",
       "      <td>83</td>\n",
       "      <td>552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>891</td>\n",
       "      <td>976</td>\n",
       "      <td>694</td>\n",
       "      <td>367</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   col0  col1  col2  col3  col4\n",
       "0   938   668   550   140   364\n",
       "1   932   978   730   919   145\n",
       "2    76   839   747   169    82\n",
       "3   624   579   770    83   552\n",
       "4   891   976   694   367   361"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's use the head function to see the first few rows of the dataframe\n",
    "new_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Quiz to practice commands for dataframes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiple_choice_quiz('pandas_operations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "zUi7SNbCiKNI"
   },
   "source": [
    "## **4) Saving numpy arrays with npy files**\n",
    "\n",
    " ```numpy``` aslo has a built in file format for you to save n-dimensional arrays and read them back in as needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "n8gptMhNiZ8h",
    "outputId": "87f30f1c-2fb2-434a-8ecb-5b08f1be91ef",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Here is an exaple for creating npy files\n",
    "import numpy as np\n",
    "\n",
    "my_array = np.ones((3,100,100))\n",
    "for i in range(3):\n",
    "  my_array[i] = my_array[i]*i\n",
    "\n",
    "print(my_array)\n",
    "\n",
    "print('')\n",
    "print('file on the disk:')\n",
    "np.save('test_save',my_array, )\n",
    "\n",
    "!head test_save.npy\n",
    "\n",
    "loaded_array = np.load('test_save.npy')\n",
    "\n",
    "print('')\n",
    "print('')\n",
    "print('loaded array:')\n",
    "print(loaded_array)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "AYysquQfRWTP"
   },
   "source": [
    "A note about memory sizes and datatypes:\n",
    "\n",
    "Numpy arrays can get large fast when you are working in multiple dimensions. When saving you can save disk space by making sure you are saving in the smallest datatype you can, keep in mind the maximum possible integers for different data types!\n",
    "\n",
    "| dtype | min | max | hex |\n",
    "| --- | --- | --- | --- |\n",
    "| int8 | -128 |127|  (-0x80, 0x7F)|\n",
    "| uint8 | 0 |255|  (0x00, 0xFF)|\n",
    "| int16 | -32768 |32767|  (-0x8000, 0x7FFF)|\n",
    "| uint16 | 0 |65535|  (0x0, 0xFFFF)|\n",
    "| int32 | -2147483648 | 2147483647 | (-0x80000000, 0x7FFFFFFF)\n",
    "| uint32 | 0 |4294967295|  (0x0, 0xFFFFFFFF)|\n",
    "| int64| -9223372036854775808 | 9223372036854775807 | (-0x8000000000000000, 0x7FFFFFFFFFFFFFFF)\n",
    "| uint64 | 0 |18446744073709551615|  (0x0, 0xFFFFFFFFFFFFFFFF)|\n",
    "\n",
    "----\n",
    "\n",
    "Note that these are maximums for **integer** values, representing floating point values is complicated and beyond the scope of this course, but 99% of the time they will use 32 or 64 bits of memory just like the larger integer data types. If you are interested check out the Wikipedia page on double precision floating points:\n",
    "https://en.wikipedia.org/wiki/Double-precision_floating-point_format\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XxDExd8lR0a8",
    "outputId": "f0a9ceb8-f3a3-4879-f2a0-187c6a200303",
    "tags": []
   },
   "outputs": [],
   "source": [
    "# here are the dtypes that numpy can handle.\n",
    "dtypes = {'int16': np.int16,\n",
    "          'int32': np.int32,\n",
    "          'float32':np.float32,\n",
    "          'float64':np.float64,\n",
    "          'bool':bool}\n",
    "\n",
    "# Let's see how much memory a 300x300x300 array would take up in different data types\n",
    "data_type = 'int16' # try with  [\"int16\",\"int32\",\"float32\", 'float64','bool']\n",
    "data_length = (100,100,200)\n",
    "dtype = dtypes[data_type]\n",
    "array = np.ones(data_length).astype(dtype)\n",
    "totalbytes = array.itemsize*array.size\n",
    "\n",
    "print(f'total mb of a {data_length} array in {data_type} is {totalbytes/1e6} MB')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "lRq21AN7lNjx"
   },
   "source": [
    "## **Questions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "id": "PlZcd6_YlNOx"
   },
   "outputs": [],
   "source": [
    "## Print the current working directory using the module os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "id": "0BYik57jwckx"
   },
   "outputs": [],
   "source": [
    "## Open a file and write a short message in text, save the file as 'example_file.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "id": "c2BM84Tpwxif"
   },
   "outputs": [],
   "source": [
    "## List out the default packages installed in your current python environment,\n",
    "## pick one, and view its docstring / documentation"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "BIOM480A5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
